{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Bioinformatics\n",
    "\n",
    "## 02 - Pairwise alignment\n",
    "\n",
    "In the first practice session, we learned how to access a file's content and how to do some basic operations with python strings. We worked with a fasta-format file that contained only one sequence. We pick up from this last point and extend our method to read multiple sequences from a fasta file. Later, we explore how a specialized library called [Biopython](https://biopython.org/) deals with sequence information. Using this library, we then compare many sequences utilizing the percentage of identity as a metric.\n",
    "\n",
    "### Multiple sequence data\n",
    "\n",
    "Most of the time, we will want to compare multiple sequences to know how they relate to each other. When these sequences are evolutionarily related, they are known as homologous sequences. Now that we are familiar with single sequence fasta files let's try to read a multiple sequence fasta file into a dictionary.\n",
    "\n",
    "To obtain the data, we will [BLAST](https://en.wikipedia.org/wiki/BLAST_(biotechnology)) our sequence and download the resulting fasta file. We will run the BLAST algorithm using the P0CG48 code inside the Uniprot page:\n",
    "\n",
    "[Uniprot BLAST page](https://www.uniprot.org/blast/)\n",
    "\n",
    "How many sequences do we have?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "with open('input/P0CG48_BLAST_sequences.fasta') as ff:\n",
    "    for l in ff:        \n",
    "        # Gather sequence lines only\n",
    "        if l.startswith('>'):\n",
    "#             count = count + 1\n",
    "            count += 1\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we put the fasta file into a dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48\n"
     ]
    }
   ],
   "source": [
    "# Define empty dictionary to store sequences in it\n",
    "sequences = {}\n",
    "\n",
    "# Define the single-sequence container as an empty string\n",
    "sequence = ''\n",
    "# Open the multiple sequences fasta file\n",
    "with open('input/P0CG48_BLAST_sequences.fasta') as ff:\n",
    "    for l in ff:        \n",
    "        # Gather sequence lines only\n",
    "        if not l.startswith('>'):\n",
    "            # Merge these lines into a single string\n",
    "            sequence += l.strip()\n",
    "        # Match fasta name lines (starting with >)\n",
    "        else:\n",
    "            # Not at the first iteration because sequence must different than ''\n",
    "            if sequence != '':\n",
    "                # Save entry in the dictionary\n",
    "                sequences[uniprot_id] = sequence\n",
    "            uniprot_id = l.split('|')[1]\n",
    "            # Empty the sequence string \n",
    "            sequence = ''\n",
    "            \n",
    "# Print the dictionary length (the number of elements in it).\n",
    "print(len(sequences.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We notice that we are short on one sequence. Why is that?\n",
    "\n",
    "After the final iteration, there is still a missing entry in the dictionary, and this is because the final iteration does not store the last iterated sequence. We only save the sequence we were gathering when we find a \">\" symbol at the beginning of the line. However, our last sequence is not followed by this symbol. This pitfall is easily managed if we tell python to store the end sequence once it has done its last iteration, i.e., after the \"for loop\" has finished."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49\n"
     ]
    }
   ],
   "source": [
    "sequences = {}\n",
    "\n",
    "sequence = ''\n",
    "with open('input/P0CG48_BLAST_sequences.fasta') as ff:\n",
    "    for l in ff:        \n",
    "        if not l.startswith('>'):\n",
    "            sequence += l.strip()\n",
    "        else:\n",
    "            if sequence != '':\n",
    "                sequences[uniprot_id] = sequence\n",
    "            uniprot_id = l.split('|')[1]\n",
    "            sequence = ''\n",
    "            \n",
    "    # Notice where this identation is located. It will save the last sequence after\n",
    "    # the \"for loop\" has finished; this is, outside the for loop.\n",
    "    sequences[uniprot_id] = sequence\n",
    "\n",
    "# Print the dictionary entries\n",
    "print(len(sequences.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'P0CG61': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGV', 'P0CG64': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGV', 'P0CG48': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGV', 'P0CG69': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGIQA', 'Q63429': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSVNTKKVKQEDTRTFLTTVSRKSPSCACSWV', 'P0CG50': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLDVEPSVTTKKVKQEDRRTFLTTVSKKSPPCACSWV', 'P0CH28': 'MQIFVKTLTGKTITLEVEPSDTIENVKGKIQEKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGVLSSPF', 'P0CG71': 'MQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLIGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGDI', 'P62976': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKQGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSNTIKKSKQEDGRTFLTTLSRKSTPCACSWA', 'P0CG66': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGV', 'P0CG68': 'MQIFVKTLTGKTITLEVEPSDTIENVKGKIQEKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGF', 'P0CG78': 'MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGN', 'P59669': 'MQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVVRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGF', 'P42740': 'MQIFVKTLTGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNNQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTITGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNLQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGM', 'Q8H159': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGF', 'P0CH05': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGDF', 'P0CH04': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGDF', 'P0CG85': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGF', 'Q39256': 'MTIQIYAKTLTEKTITLDVETSDSIHNVKAKIQNKEGIPLDQQRLIFAGKQLEDGLTLADYNIQKESTLHLVLRLRGGMQIFVQTLTGKTITLEVKSSDTIDNVKAKIQDKEGILPRQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLCGGMQIFVSTFSGKNFTSDTLTLKVESSDTIENVKAKIQDREGLRPDHQRLIFHGEELFTEDNRTLADYGIRNRSTLCLALRLRGDMYIFVKNLPYNSFTGENFILEVESSDTIDNVKAKLQDKERIPMDLHRLIFAGKPLEGGRTLAHYNIQKGSTLYLVTRFRCGMQIFVKTLTRKRINLEVESWDTIENVKAMVQDKEGIQPQPNLQRLIFLGKELKDGCTLADYSIQKESTLHLVLGMQIFVKLFGGKIITLEVLSSDTIKSVKAKIQDKVGSPPDQQILLFRGGQLQDGRTLGDYNIRNESTLHLFFHIRHGMQIFVKTFSFSGETPTCKTITLEVESSDTIDNVKVKIQHKVGIPLDRQRLIFGGRVLVGSRTLLDYNIQKGSTIHQLFLQRGGMQIFIKTLTGKTIILEVESSDTIANVKEKIQVKEGIKPDQQMLIFFGQQLEDGVTLGDYDIHKKSTLYLVLRLRQRRYDF', 'P42739': 'IPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVQSSDTVENVKSKIQDKEGIPPDQQRLIFAGKQLEDGLTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTVENVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTVENVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTVENVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTVENVKSKIQDKEGIPPDQQRIIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGAF', 'P0CG82': 'MQIFVKTLTGKTITLDVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLDVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLDVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLDVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLDVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGQ', 'P0CG75': 'MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGI', 'P0CG79': 'MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGL', 'P0CG72': 'MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGLF', 'P0CH32': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGSF', 'P69322': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGF', 'P0CG63': 'MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGN', 'Q58G87': 'MQIFVKTLTGKTITLEVEFSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGL', 'P0CG81': 'MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVECSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGN', 'P0CG76': 'MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGN', 'P0CG84': 'VKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGF', 'P0CG54': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGVYASPIF', 'P0CG51': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGY', 'P0CG49': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGY', 'P0CG62': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGY', 'P0CG53': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGC', 'P0CG55': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRMRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGC', 'Q8MKD1': 'MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRFIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGC', 'P69315': 'IPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFLKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGI', 'Q1EC66': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGSF', 'Q3E7T8': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGF', 'P69325': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGF', 'P69309': 'MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGQ', 'P0CG88': 'MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGN', 'P0CG80': 'MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGN', 'P0CG74': 'MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGF', 'P0CH27': 'MQIFVKTLTGKTIALEVESSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTIALEVESSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTIALEVESSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTIALEVESSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGVMEPTLEALAKKYNWEKKVCRRCYARLPVRASNCRKKACGHCSNLRMKKKLR', 'P0CG70': 'MQIFVKTLTGKTITLEVESSDTIDNVKQKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKQKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKQKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKQKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGQ', 'Q9FHQ6': 'MSMQIHAKTLTEKTITIDVVSSDTINNVKAKIQDIEGIPLDQQRLIFSGKLLDDGRTLADYSIQKDSILHLALRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGVPPDQQRLIFAGKQLDDGRTLADYNIQKESTLHLVLRLRGGMQIFVRTLTRKTIALEVESSDTTDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLCGGMQIFVNTLTGKTITLEVESSDTIDNVKAKIQDKERIQPDQQRLIFAGEQLEDGYYTLADYNIQKESTLHLVLRLRGGECFGFIFLFLLCFNS'}\n"
     ]
    }
   ],
   "source": [
    "# Let's print our dictionary to see how it looks like.\n",
    "print(sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Declaring python functions\n",
    "\n",
    "Note that the previous code is sufficiently generic to be applied to any fasta file. You can reuse the code as many times as you want if you wrap it as a Python function. If the functions you write are generalizable, they can be applied in many different contexts, saving us a lot of typing!\n",
    "\n",
    "We will convert the above code into a function. This function will get as input the path to the fasta file, and it will return a dictionary containing the sequences. Also, the function must have proper documentation. The documentation should include the definition and usage of the function. In this way, the function can be understood and used by others.\n",
    "\n",
    "To declare a function, we begin with the word \"def,\" followed by the function's name, inputs, and keywords. The function ends when it encounters a return statement. This statement allows you to store the function's output into a variable after its execution.\n",
    "\n",
    "```\n",
    "def myfunction(arguments, keywords=None):\n",
    "    \"\"\"\n",
    "    Here goes the documentation\n",
    "    \"\"\"\n",
    "    \n",
    "    Here goes the code\n",
    "    \n",
    "    Here goes the return statement\n",
    "    \n",
    "```  \n",
    "     \n",
    "Let's declare a function to read fasta files as dictionaries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFasta(input_fasta):\n",
    "    \"\"\"\n",
    "    Reads a fasta file and returns the sequence data as a dictionary.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    input_fasta : str\n",
    "        Path to the input fasta file\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    sequences : dict\n",
    "        Dictionary containing the IDs and squences in the fasta file.\n",
    "    \"\"\"\n",
    "    sequences = {}\n",
    "    sequence = ''\n",
    "    with open(input_fasta) as ff:\n",
    "        for l in ff:        \n",
    "            if not l.startswith('>'):\n",
    "                sequence += l.strip()\n",
    "            else:\n",
    "                if sequence != '':\n",
    "                    sequences[uniprot_id] = sequence\n",
    "                uniprot_id = l.split('|')[1]\n",
    "                sequence = ''\n",
    "        sequences[uniprot_id] = sequence\n",
    "    \n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['P0CG61', 'P0CG64', 'P0CG48', 'P0CG69', 'Q63429', 'P0CG50', 'P0CH28', 'P0CG71', 'P62976', 'P0CG66', 'P0CG68', 'P0CG78', 'P59669', 'P42740', 'Q8H159', 'P0CH05', 'P0CH04', 'P0CG85', 'Q39256', 'P42739', 'P0CG82', 'P0CG75', 'P0CG79', 'P0CG72', 'P0CH32', 'P69322', 'P0CG63', 'Q58G87', 'P0CG81', 'P0CG76', 'P0CG84', 'P0CG54', 'P0CG51', 'P0CG49', 'P0CG62', 'P0CG53', 'P0CG55', 'Q8MKD1', 'P69315', 'Q1EC66', 'Q3E7T8', 'P69325', 'P69309', 'P0CG88', 'P0CG80', 'P0CG74', 'P0CH27', 'P0CG70', 'Q9FHQ6'])\n"
     ]
    }
   ],
   "source": [
    "sequences = readFasta('input/P0CG48_BLAST_sequences.fasta')\n",
    "print(sequences.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This syntax is a compelling way of programming. You can reuse any code previously made and put it to fair use in different contexts by making functions.\n",
    "\n",
    "To get help for a function, you just call:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function readFasta in module __main__:\n",
      "\n",
      "readFasta(input_fasta)\n",
      "    Reads a fasta file and returns the sequence data as a dictionary.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    input_fasta : str\n",
      "        Path to the input fasta file\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    sequences : dict\n",
      "        Dictionary containing the IDs and squences in the fasta file.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(readFasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This prints out the documentation of the function. Acceptable documentation should give you all you need to understand how the function is used. To get specific details of a function, you need to look at its code and comments within.\n",
    "\n",
    "### Read sequences with Biopython \n",
    "\n",
    "Now that you are aware of how files can be read and parsed by python. We will move forward and use a specialized library to read, store and analyze our data. It uses more sophisticated functions to read and parse fasta files and other file formats relevant for storing biological data.\n",
    "\n",
    "Let's read our blast fasta file with Biopython. We start by importing Biopython and all other libraries we will need."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Biopython has a SeqIO interface that allows us to read fasta formatted files. We use the method parse to read the data into a generator, a particular type of iterator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: sp|P0CG61|UBC_PONPY\n",
      "Name: sp|P0CG61|UBC_PONPY\n",
      "Description: sp|P0CG61|UBC_PONPY Polyubiquitin-C OS=Pongo pygmaeus OX=9600 GN=UBC PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGV', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG64|UBC_PANTR\n",
      "Name: sp|P0CG64|UBC_PANTR\n",
      "Description: sp|P0CG64|UBC_PANTR Polyubiquitin-C OS=Pan troglodytes OX=9598 GN=UBC PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGV', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG48|UBC_HUMAN\n",
      "Name: sp|P0CG48|UBC_HUMAN\n",
      "Description: sp|P0CG48|UBC_HUMAN Polyubiquitin-C OS=Homo sapiens OX=9606 GN=UBC PE=1 SV=3\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGV', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG69|UBIQP_DROME\n",
      "Name: sp|P0CG69|UBIQP_DROME\n",
      "Description: sp|P0CG69|UBIQP_DROME Polyubiquitin OS=Drosophila melanogaster OX=7227 GN=Ubi-p63E PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...IQA', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q63429|UBC_RAT\n",
      "Name: sp|Q63429|UBC_RAT\n",
      "Description: sp|Q63429|UBC_RAT Polyubiquitin-C OS=Rattus norvegicus OX=10116 GN=Ubc PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...SWV', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG50|UBC_MOUSE\n",
      "Name: sp|P0CG50|UBC_MOUSE\n",
      "Description: sp|P0CG50|UBC_MOUSE Polyubiquitin-C OS=Mus musculus OX=10090 GN=Ubc PE=1 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...SWV', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CH28|UBC_BOVIN\n",
      "Name: sp|P0CH28|UBC_BOVIN\n",
      "Description: sp|P0CH28|UBC_BOVIN Polyubiquitin-C OS=Bos taurus OX=9913 GN=UBC PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKGKIQEKEGIPPDQQRLIFAGKQLEDGR...SPF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG71|UBIQ1_CAEEL\n",
      "Name: sp|P0CG71|UBIQ1_CAEEL\n",
      "Description: sp|P0CG71|UBIQ1_CAEEL Polyubiquitin-A OS=Caenorhabditis elegans OX=6239 GN=ubq-1 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GDI', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P62976|UBIQP_CRIGR\n",
      "Name: sp|P62976|UBIQP_CRIGR\n",
      "Description: sp|P62976|UBIQP_CRIGR Polyubiquitin OS=Cricetulus griseus OX=10029 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...SWA', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG66|UBC_GORGO\n",
      "Name: sp|P0CG66|UBC_GORGO\n",
      "Description: sp|P0CG66|UBC_GORGO Polyubiquitin-C OS=Gorilla gorilla gorilla OX=9595 GN=UBC PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGV', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG68|UBC_PIG\n",
      "Name: sp|P0CG68|UBC_PIG\n",
      "Description: sp|P0CG68|UBC_PIG Polyubiquitin-C OS=Sus scrofa OX=9823 GN=UBC PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKGKIQEKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG78|UBIQF_DICDI\n",
      "Name: sp|P0CG78|UBIQF_DICDI\n",
      "Description: sp|P0CG78|UBIQF_DICDI Polyubiquitin-F OS=Dictyostelium discoideum OX=44689 GN=ubqF PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGN', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P59669|UBIQP_GEOCY\n",
      "Name: sp|P59669|UBIQP_GEOCY\n",
      "Description: sp|P59669|UBIQP_GEOCY Polyubiquitin OS=Geodia cydonium OX=6047 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P42740|UBIQP_AGLNE\n",
      "Name: sp|P42740|UBIQP_AGLNE\n",
      "Description: sp|P42740|UBIQP_AGLNE Polyubiquitin OS=Aglaothamnion neglectum OX=2765 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIENVKTKIQDKEGIPPDQQRLIFAGKQLEDGR...GGM', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q8H159|UBQ10_ARATH\n",
      "Name: sp|Q8H159|UBQ10_ARATH\n",
      "Description: sp|Q8H159|UBQ10_ARATH Polyubiquitin 10 OS=Arabidopsis thaliana OX=3702 GN=UBQ10 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CH05|UBI2P_PETCR\n",
      "Name: sp|P0CH05|UBI2P_PETCR\n",
      "Description: sp|P0CH05|UBI2P_PETCR Polyubiquitin OS=Petroselinum crispum OX=4043 GN=PCUBI4-2 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GDF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CH04|UBI1P_PETCR\n",
      "Name: sp|P0CH04|UBI1P_PETCR\n",
      "Description: sp|P0CH04|UBI1P_PETCR Polyubiquitin OS=Petroselinum crispum OX=4043 GN=PCUBI4-1 PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GDF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG85|UBI1P_NICSY\n",
      "Name: sp|P0CG85|UBI1P_NICSY\n",
      "Description: sp|P0CG85|UBI1P_NICSY Polyubiquitin OS=Nicotiana sylvestris OX=4096 GN=UBI11 PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q39256|UBQ8_ARATH\n",
      "Name: sp|Q39256|UBQ8_ARATH\n",
      "Description: sp|Q39256|UBQ8_ARATH Polyubiquitin 8 OS=Arabidopsis thaliana OX=3702 GN=UBQ8 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MTIQIYAKTLTEKTITLDVETSDSIHNVKAKIQNKEGIPLDQQRLIFAGKQLED...YDF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P42739|UBIQP_ACEPE\n",
      "Name: sp|P42739|UBIQP_ACEPE\n",
      "Description: sp|P42739|UBIQP_ACEPE Polyubiquitin (Fragment) OS=Acetabularia peniculus OX=35862 PE=3 SV=2\n",
      "Number of features: 0\n",
      "Seq('IPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFVKTLTGKTI...GAF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG82|UBIQP_TETPY\n",
      "Name: sp|P0CG82|UBIQP_TETPY\n",
      "Description: sp|P0CG82|UBIQP_TETPY Polyubiquitin OS=Tetrahymena pyriformis OX=5908 GN=TU20 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLDVEASDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGQ', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG75|UBI4P_KLULA\n",
      "Name: sp|P0CG75|UBI4P_KLULA\n",
      "Description: sp|P0CG75|UBI4P_KLULA Polyubiquitin OS=Kluyveromyces lactis (strain ATCC 8585 / CBS 2359 / DSM 70799 / NBRC 1267 / NRRL Y-1140 / WM37) OX=284590 GN=ubi4 PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGR...GGI', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG79|UBIQG_DICDI\n",
      "Name: sp|P0CG79|UBIQG_DICDI\n",
      "Description: sp|P0CG79|UBIQG_DICDI Polyubiquitin-G OS=Dictyostelium discoideum OX=44689 GN=ubqG PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGL', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG72|UBI4P_SCHPO\n",
      "Name: sp|P0CG72|UBI4P_SCHPO\n",
      "Description: sp|P0CG72|UBI4P_SCHPO Polyubiquitin OS=Schizosaccharomyces pombe (strain 972 / ATCC 24843) OX=284812 GN=ubi4 PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGR...GLF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CH32|UBQ4_ARATH\n",
      "Name: sp|P0CH32|UBQ4_ARATH\n",
      "Description: sp|P0CH32|UBQ4_ARATH Polyubiquitin 4 OS=Arabidopsis thaliana OX=3702 GN=UBQ4 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GSF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P69322|UBIQP_PEA\n",
      "Name: sp|P69322|UBIQP_PEA\n",
      "Description: sp|P69322|UBIQP_PEA Polyubiquitin OS=Pisum sativum OX=3888 GN=PU1 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG63|UBI4P_YEAST\n",
      "Name: sp|P0CG63|UBI4P_YEAST\n",
      "Description: sp|P0CG63|UBI4P_YEAST Polyubiquitin OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) OX=559292 GN=UBI4 PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGR...GGN', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q58G87|UBQ3_ORYSJ\n",
      "Name: sp|Q58G87|UBQ3_ORYSJ\n",
      "Description: sp|Q58G87|UBQ3_ORYSJ Polyubiquitin 3 OS=Oryza sativa subsp. japonica OX=39947 GN=UBQ3 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEFSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGL', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG81|UBIQH_DICDI\n",
      "Name: sp|P0CG81|UBIQH_DICDI\n",
      "Description: sp|P0CG81|UBIQH_DICDI Polyubiquitin-H OS=Dictyostelium discoideum OX=44689 GN=ubqH PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGN', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG76|UBIQA_DICDI\n",
      "Name: sp|P0CG76|UBIQA_DICDI\n",
      "Description: sp|P0CG76|UBIQA_DICDI Polyubiquitin-A OS=Dictyostelium discoideum OX=44689 GN=ubqA PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGN', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG84|UBI4P_NICSY\n",
      "Name: sp|P0CG84|UBI4P_NICSY\n",
      "Description: sp|P0CG84|UBI4P_NICSY Polyubiquitin (Fragment) OS=Nicotiana sylvestris OX=4096 GN=UBI4 PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('VKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLAD...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG54|UBB_CAVPO\n",
      "Name: sp|P0CG54|UBB_CAVPO\n",
      "Description: sp|P0CG54|UBB_CAVPO Polyubiquitin-B OS=Cavia porcellus OX=10141 GN=UBB PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...PIF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG51|UBB_RAT\n",
      "Name: sp|P0CG51|UBB_RAT\n",
      "Description: sp|P0CG51|UBB_RAT Polyubiquitin-B OS=Rattus norvegicus OX=10116 GN=Ubb PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGY', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG49|UBB_MOUSE\n",
      "Name: sp|P0CG49|UBB_MOUSE\n",
      "Description: sp|P0CG49|UBB_MOUSE Polyubiquitin-B OS=Mus musculus OX=10090 GN=Ubb PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGY', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG62|UBB_CHICK\n",
      "Name: sp|P0CG62|UBB_CHICK\n",
      "Description: sp|P0CG62|UBB_CHICK Polyubiquitin-B OS=Gallus gallus OX=9031 GN=UBB PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGY', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG53|UBB_BOVIN\n",
      "Name: sp|P0CG53|UBB_BOVIN\n",
      "Description: sp|P0CG53|UBB_BOVIN Polyubiquitin-B OS=Bos taurus OX=9913 GN=UBB PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGC', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG55|UBB_SHEEP\n",
      "Name: sp|P0CG55|UBB_SHEEP\n",
      "Description: sp|P0CG55|UBB_SHEEP Polyubiquitin-B OS=Ovis aries OX=9940 GN=UBB PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGC', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q8MKD1|UBB_HORSE\n",
      "Name: sp|Q8MKD1|UBB_HORSE\n",
      "Description: sp|Q8MKD1|UBB_HORSE Polyubiquitin-B OS=Equus caballus OX=9796 GN=UBB PE=2 SV=3\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEPSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGC', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P69315|UBIQP_LINUS\n",
      "Name: sp|P69315|UBIQP_LINUS\n",
      "Description: sp|P69315|UBIQP_LINUS Polyubiquitin (Fragment) OS=Linum usitatissimum OX=4006 PE=3 SV=2\n",
      "Number of features: 0\n",
      "Seq('IPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLRGGMQIFLKTLTGKTI...EGI', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q1EC66|UBQ3_ARATH\n",
      "Name: sp|Q1EC66|UBQ3_ARATH\n",
      "Description: sp|Q1EC66|UBQ3_ARATH Polyubiquitin 3 OS=Arabidopsis thaliana OX=3702 GN=UBQ3 PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GSF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q3E7T8|UBQ14_ARATH\n",
      "Name: sp|Q3E7T8|UBQ14_ARATH\n",
      "Description: sp|Q3E7T8|UBQ14_ARATH Polyubiquitin 14 OS=Arabidopsis thaliana OX=3702 GN=UBQ14 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P69325|UBIQP_SOYBN\n",
      "Name: sp|P69325|UBIQP_SOYBN\n",
      "Description: sp|P69325|UBIQP_SOYBN Polyubiquitin OS=Glycine max OX=3847 GN=SUBI-1 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P69309|UBIQP_AVEFA\n",
      "Name: sp|P69309|UBIQP_AVEFA\n",
      "Description: sp|P69309|UBIQP_AVEFA Polyubiquitin OS=Avena fatua OX=4499 PE=2 SV=2\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGQ', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG88|UBIQJ_DICDI\n",
      "Name: sp|P0CG88|UBIQJ_DICDI\n",
      "Description: sp|P0CG88|UBIQJ_DICDI Polyubiquitin-J OS=Dictyostelium discoideum OX=44689 GN=ubqJ PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGN', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG80|UBIQI_DICDI\n",
      "Name: sp|P0CG80|UBIQI_DICDI\n",
      "Description: sp|P0CG80|UBIQI_DICDI Polyubiquitin-I OS=Dictyostelium discoideum OX=44689 GN=ubqI PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVEGSDNIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...GGN', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG74|UBI4P_CANAX\n",
      "Name: sp|P0CG74|UBI4P_CANAX\n",
      "Description: sp|P0CG74|UBI4P_CANAX Polyubiquitin OS=Candida albicans OX=5476 GN=UBI4 PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGR...GGF', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CH27|RL402_TRYCR\n",
      "Name: sp|P0CH27|RL402_TRYCR\n",
      "Description: sp|P0CH27|RL402_TRYCR Ubiquitin-60S ribosomal protein L40 OS=Trypanosoma cruzi OX=5693 PE=2 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTIALEVESSDTIENVKAKIQDKEGIPPDQQRLIFAGKQLEDGR...KLR', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|P0CG70|UBI4P_NEUCR\n",
      "Name: sp|P0CG70|UBI4P_NEUCR\n",
      "Description: sp|P0CG70|UBI4P_NEUCR Polyubiquitin OS=Neurospora crassa (strain ATCC 24698 / 74-OR23-1A / CBS 708.71 / DSM 1257 / FGSC 987) OX=367110 GN=ubi PE=1 SV=1\n",
      "Number of features: 0\n",
      "Seq('MQIFVKTLTGKTITLEVESSDTIDNVKQKIQDKEGIPPDQQRLIFAGKQLEDGR...GGQ', SingleLetterAlphabet())\n",
      "\n",
      "ID: sp|Q9FHQ6|UBQ9_ARATH\n",
      "Name: sp|Q9FHQ6|UBQ9_ARATH\n",
      "Description: sp|Q9FHQ6|UBQ9_ARATH Polyubiquitin 9 OS=Arabidopsis thaliana OX=3702 GN=UBQ9 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('MSMQIHAKTLTEKTITIDVVSSDTINNVKAKIQDIEGIPLDQQRLIFSGKLLDD...FNS', SingleLetterAlphabet())\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Read fasta file with Biopython's SeqIO interface\n",
    "sequences = SeqIO.parse('input/P0CG48_BLAST_sequences.fasta', 'fasta')\n",
    "\n",
    "#We print the sequences contained in the iterator with a for loop\n",
    "for sequence in sequences:\n",
    "    print(sequence)\n",
    "    print() # Empty print to separate each iteration's print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we use the help() method to understand how Biopython represents sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on SeqRecord in module Bio.SeqRecord object:\n",
      "\n",
      "class SeqRecord(builtins.object)\n",
      " |  A SeqRecord object holds a sequence and information about it.\n",
      " |  \n",
      " |  Main attributes:\n",
      " |   - id          - Identifier such as a locus tag (string)\n",
      " |   - seq         - The sequence itself (Seq object or similar)\n",
      " |  \n",
      " |  Additional attributes:\n",
      " |   - name        - Sequence name, e.g. gene name (string)\n",
      " |   - description - Additional text (string)\n",
      " |   - dbxrefs     - List of database cross references (list of strings)\n",
      " |   - features    - Any (sub)features defined (list of SeqFeature objects)\n",
      " |   - annotations - Further information about the whole sequence (dictionary).\n",
      " |     Most entries are strings, or lists of strings.\n",
      " |   - letter_annotations - Per letter/symbol annotation (restricted\n",
      " |     dictionary). This holds Python sequences (lists, strings\n",
      " |     or tuples) whose length matches that of the sequence.\n",
      " |     A typical use would be to hold a list of integers\n",
      " |     representing sequencing quality scores, or a string\n",
      " |     representing the secondary structure.\n",
      " |  \n",
      " |  You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |  SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |  objects directly (see the __init__ method for further details):\n",
      " |  \n",
      " |  >>> from Bio.Seq import Seq\n",
      " |  >>> from Bio.SeqRecord import SeqRecord\n",
      " |  >>> from Bio.Alphabet import IUPAC\n",
      " |  >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |  ...                         IUPAC.protein),\n",
      " |  ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |  ...                    description=\"toxic membrane protein\")\n",
      " |  >>> print(record)\n",
      " |  ID: YP_025292.1\n",
      " |  Name: HokC\n",
      " |  Description: toxic membrane protein\n",
      " |  Number of features: 0\n",
      " |  Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |  \n",
      " |  If you want to save SeqRecord objects to a sequence file, use Bio.SeqIO\n",
      " |  for this.  For the special case where you want the SeqRecord turned into\n",
      " |  a string in a particular file format there is a format method which uses\n",
      " |  Bio.SeqIO internally:\n",
      " |  \n",
      " |  >>> print(record.format(\"fasta\"))\n",
      " |  >YP_025292.1 toxic membrane protein\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also do things like slicing a SeqRecord, checking its length, etc\n",
      " |  \n",
      " |  >>> len(record)\n",
      " |  44\n",
      " |  >>> edited = record[:10] + record[11:]\n",
      " |  >>> print(edited.seq)\n",
      " |  MKQHKAMIVAIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  >>> print(record.seq)\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      The other sequence can be a SeqRecord object, a Seq object (or\n",
      " |      similar, e.g. a MutableSeq) or a plain Python string. If you add\n",
      " |      a plain string or a Seq (like) object, the new SeqRecord will simply\n",
      " |      have this appended to the existing data. However, any per letter\n",
      " |      annotation will be lost:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = record + \"ACT\"\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNNACT\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |      \n",
      " |      The new record will attempt to combine the annotation, but for any\n",
      " |      ambiguities (e.g. different names) it defaults to omitting that\n",
      " |      annotation.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      \n",
      " |      Now let's cut the plasmid into two pieces, and join them back up the\n",
      " |      other way round (i.e. shift the starting point on this plasmid, have\n",
      " |      a look at the annotated features in the original file to see why this\n",
      " |      particular split point might make sense):\n",
      " |      \n",
      " |      >>> left = plasmid[:3765]\n",
      " |      >>> right = plasmid[3765:]\n",
      " |      >>> new = right + left\n",
      " |      >>> print(\"%s %i\" % (new.id, len(new)))\n",
      " |      pBAD30 4923\n",
      " |      >>> str(new.seq) == str(right.seq + left.seq)\n",
      " |      True\n",
      " |      >>> len(new.features) == len(left.features) + len(right.features)\n",
      " |      True\n",
      " |      \n",
      " |      When we add the left and right SeqRecord objects, their annotation\n",
      " |      is all consistent, so it is all conserved in the new SeqRecord:\n",
      " |      \n",
      " |      >>> new.id == left.id == right.id == plasmid.id\n",
      " |      True\n",
      " |      >>> new.name == left.name == right.name == plasmid.name\n",
      " |      True\n",
      " |      >>> new.description == plasmid.description\n",
      " |      True\n",
      " |      >>> new.annotations == left.annotations == right.annotations\n",
      " |      True\n",
      " |      >>> new.letter_annotations == plasmid.letter_annotations\n",
      " |      True\n",
      " |      >>> new.dbxrefs == left.dbxrefs == right.dbxrefs\n",
      " |      True\n",
      " |      \n",
      " |      However, we should point out that when we sliced the SeqRecord,\n",
      " |      any annotations dictionary or dbxrefs list entries were lost.\n",
      " |      You can explicitly copy them like this:\n",
      " |      \n",
      " |      >>> new.annotations = plasmid.annotations.copy()\n",
      " |      >>> new.dbxrefs = plasmid.dbxrefs[:]\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Boolean value of an instance of this class (True).\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a SeqRecord always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, a Seq object will evaluate to False if it\n",
      " |      has a zero length sequence.\n",
      " |      \n",
      " |      WARNING: The SeqRecord may in future evaluate to False when its\n",
      " |      sequence is of zero length (in order to better match the Seq\n",
      " |      object behaviour)!\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implement the 'in' keyword, searches the sequence.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> \"GAATTC\" in record\n",
      " |      False\n",
      " |      >>> \"AAA\" in record\n",
      " |      True\n",
      " |      \n",
      " |      This essentially acts as a proxy for using \"in\" on the sequence:\n",
      " |      \n",
      " |      >>> \"GAATTC\" in record.seq\n",
      " |      False\n",
      " |      >>> \"AAA\" in record.seq\n",
      " |      True\n",
      " |      \n",
      " |      Note that you can also use Seq objects as the query,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> Seq(\"AAA\") in record\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\", generic_dna) in record\n",
      " |      True\n",
      " |      \n",
      " |      See also the Seq object's __contains__ method.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Define the equal-to operand (not implemented).\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      This method supports the python format() function added in\n",
      " |      Python 2.6/3.0.  The format_spec should be a lower case string\n",
      " |      supported by Bio.SeqIO as an output file format. See also the\n",
      " |      SeqRecord's format() method.\n",
      " |      \n",
      " |      Under Python 3 binary formats raise a ValueError, while on\n",
      " |      Python 2 this will work with a deprecation warning.\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Define the greater-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a sub-sequence or an individual letter.\n",
      " |      \n",
      " |      Slicing, e.g. my_record[5:10], returns a new SeqRecord for\n",
      " |      that sub-sequence with some annotation preserved as follows:\n",
      " |      \n",
      " |      * The name, id and description are kept as-is.\n",
      " |      * Any per-letter-annotations are sliced to match the requested\n",
      " |        sub-sequence.\n",
      " |      * Unless a stride is used, all those features which fall fully\n",
      " |        within the subsequence are included (with their locations\n",
      " |        adjusted accordingly). If you want to preserve any truncated\n",
      " |        features (e.g. GenBank/EMBL source features), you must\n",
      " |        explicitly add them to the new SeqRecord yourself.\n",
      " |      * The annotations dictionary and the dbxrefs list are not used\n",
      " |        for the new SeqRecord, as in general they may not apply to the\n",
      " |        subsequence. If you want to preserve them, you must explicitly\n",
      " |        copy them to the new SeqRecord yourself.\n",
      " |      \n",
      " |      Using an integer index, e.g. my_record[5] is shorthand for\n",
      " |      extracting that letter from the sequence, my_record.seq[5].\n",
      " |      \n",
      " |      For example, consider this short protein and its secondary\n",
      " |      structure as encoded by the PDB (e.g. H for alpha helices),\n",
      " |      plus a simple feature for its histidine self phosphorylation\n",
      " |      site:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> rec = SeqRecord(Seq(\"MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLAT\"\n",
      " |      ...                     \"EMMSEQDGYLAESINKDIEECNAIIEQFIDYLR\",\n",
      " |      ...                     IUPAC.protein),\n",
      " |      ...                 id=\"1JOY\", name=\"EnvZ\",\n",
      " |      ...                 description=\"Homodimeric domain of EnvZ from E. coli\")\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"] = \"  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \"\n",
      " |      >>> rec.features.append(SeqFeature(FeatureLocation(20, 21),\n",
      " |      ...                     type = \"Site\"))\n",
      " |      \n",
      " |      Now let's have a quick look at the full record,\n",
      " |      \n",
      " |      >>> print(rec)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR', IUPACProtein())\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"]\n",
      " |      '  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  '\n",
      " |      >>> print(rec.features[0].location)\n",
      " |      [20:21]\n",
      " |      \n",
      " |      Now let's take a sub sequence, here chosen as the first (fractured)\n",
      " |      alpha helix which includes the histidine phosphorylation site:\n",
      " |      \n",
      " |      >>> sub = rec[11:41]\n",
      " |      >>> print(sub)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('RTLLMAGVSHDLRTPLTRIRLATEMMSEQD', IUPACProtein())\n",
      " |      >>> sub.letter_annotations[\"secondary_structure\"]\n",
      " |      'HHHHHTTTHHHHHHHHHHHHHHHHHHHHHH'\n",
      " |      >>> print(sub.features[0].location)\n",
      " |      [9:10]\n",
      " |      \n",
      " |      You can also of course omit the start or end values, for\n",
      " |      example to get the first ten letters only:\n",
      " |      \n",
      " |      >>> print(rec[:10])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLAD', IUPACProtein())\n",
      " |      \n",
      " |      Or for the last ten letters:\n",
      " |      \n",
      " |      >>> print(rec[-10:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('IIEQFIDYLR', IUPACProtein())\n",
      " |      \n",
      " |      If you omit both, then you get a copy of the original record (although\n",
      " |      lacking the annotations and dbxrefs):\n",
      " |      \n",
      " |      >>> print(rec[:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR', IUPACProtein())\n",
      " |      \n",
      " |      Finally, indexing with a simple integer is shorthand for pulling out\n",
      " |      that letter from the sequence directly:\n",
      " |      \n",
      " |      >>> rec[5]\n",
      " |      'K'\n",
      " |      >>> rec.seq[5]\n",
      " |      'K'\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Define the greater-than operand (not implemented).\n",
      " |  \n",
      " |  __init__(self, seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |      Create a SeqRecord.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - seq         - Sequence, required (Seq, MutableSeq or UnknownSeq)\n",
      " |       - id          - Sequence identifier, recommended (string)\n",
      " |       - name        - Sequence name, optional (string)\n",
      " |       - description - Sequence description, optional (string)\n",
      " |       - dbxrefs     - Database cross references, optional (list of strings)\n",
      " |       - features    - Any (sub)features, optional (list of SeqFeature objects)\n",
      " |       - annotations - Dictionary of annotations for the whole sequence\n",
      " |       - letter_annotations - Dictionary of per-letter-annotations, values\n",
      " |         should be strings, list or tuples of the same length as the full\n",
      " |         sequence.\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |      objects directly.\n",
      " |      \n",
      " |      Note that while an id is optional, we strongly recommend you supply a\n",
      " |      unique id string for each record.  This is especially important\n",
      " |      if you wish to write your sequences to a file.\n",
      " |      \n",
      " |      If you don't have the actual sequence, but you do know its length,\n",
      " |      then using the UnknownSeq object from Bio.Seq is appropriate.\n",
      " |      \n",
      " |      You can create a 'blank' SeqRecord object, and then populate the\n",
      " |      attributes later.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the letters in the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a protein FASTA file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/loveliesbleeding.pro\", \"fasta\")\n",
      " |      >>> for amino in record:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      This is just a shortcut for iterating over the sequence directly:\n",
      " |      \n",
      " |      >>> for amino in record.seq:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      Note that this does not facilitate iteration together with any\n",
      " |      per-letter-annotation.  However, you can achieve that using the\n",
      " |      python zip function on the record (or its sequence) and the relevant\n",
      " |      per-letter-annotation:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> rec = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(rec.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> for nuc, qual in zip(rec, rec.letter_annotations[\"solexa_quality\"]):\n",
      " |      ...     if qual > 35:\n",
      " |      ...         print(\"%s %i\" % (nuc, qual))\n",
      " |      A 40\n",
      " |      C 39\n",
      " |      G 38\n",
      " |      T 37\n",
      " |      A 36\n",
      " |      \n",
      " |      You may agree that using zip(rec.seq, ...) is more explicit than using\n",
      " |      zip(rec, ...) as shown above.\n",
      " |  \n",
      " |  __le___(self, other)\n",
      " |      Define the less-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a FASTA nucleotide file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> len(record)\n",
      " |      309\n",
      " |      >>> len(record.seq)\n",
      " |      309\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Define the less-than operand (not implemented).\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Define the not-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __nonzero__ = __bool__(self)\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add another sequence or string to this sequence (from the left).\n",
      " |      \n",
      " |      This method handles adding a Seq object (or similar, e.g. MutableSeq)\n",
      " |      or a plain Python string (on the left) to a SeqRecord (on the right).\n",
      " |      See the __add__ method for more details, but for example:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = \"ACT\" + record\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a concise summary of the record for debugging (string).\n",
      " |      \n",
      " |      The python built in function repr works by calling the object's ___repr__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> rec = SeqRecord(Seq(\"MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKAT\"\n",
      " |      ...                    +\"GEMKEQTEWHRVVLFGKLAEVASEYLRKGSQVYIEGQLRTRKWTDQ\"\n",
      " |      ...                    +\"SGQDRYTTEVVVNVGGTMQMLGGRQGGGAPAGGNIGGGQPQGGWGQ\"\n",
      " |      ...                    +\"PQQPQGGNQFSGGAQSRPQQSAPAAPSNEPPMDFDDDIPF\",\n",
      " |      ...                    generic_protein),\n",
      " |      ...                 id=\"NP_418483.1\", name=\"b4059\",\n",
      " |      ...                 description=\"ssDNA-binding protein\",\n",
      " |      ...                 dbxrefs=[\"ASAP:13298\", \"GI:16131885\", \"GeneID:948570\"])\n",
      " |      >>> print(repr(rec))\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF', ProteinAlphabet()), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      At the python prompt you can also use this shorthand:\n",
      " |      \n",
      " |      >>> rec\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF', ProteinAlphabet()), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      Note that long sequences are shown truncated. Also note that any\n",
      " |      annotations, letter_annotations and features are not shown (as they\n",
      " |      would lead to a very long string).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a human readable summary of the record and its annotation (string).\n",
      " |      \n",
      " |      The python built in function str works by calling the object's ___str__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...                         IUPAC.protein),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein, small\")\n",
      " |      >>> print(str(record))\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      \n",
      " |      In this example you don't actually need to call str explicity, as the\n",
      " |      print command does this automatically:\n",
      " |      \n",
      " |      >>> print(record)\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      \n",
      " |      Note that long sequences are shown truncated.\n",
      " |  \n",
      " |  format(self, format)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.SeqIO, which is used to turn the SeqRecord into a\n",
      " |      string.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...                         IUPAC.protein),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      >>> record.format(\"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      The python print command automatically appends a new line, meaning\n",
      " |      in this example a blank line is shown.  If you look at the string\n",
      " |      representation you can see there is a trailing new line (shown as\n",
      " |      slash n) which is important when writing to a file or if\n",
      " |      concatenating multiple sequence strings together.\n",
      " |      \n",
      " |      Note that this method will NOT work on every possible file format\n",
      " |      supported by Bio.SeqIO (e.g. some are for multiple sequences only,\n",
      " |      and binary formats are not supported).\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Return a copy of the record with a lower case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/aster.pro\", \"fasta\")\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      GGHVNPAVTFGAFVGGNITLLRGIVYIIAQLLGSTVACLLLKFVTNDMAVGVFSLSAGVG\n",
      " |      VTNALVFEIVMTFGLVYTVYATAIDPKKGSLGTIAPIAIGFIVGANI\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(record.lower().format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      gghvnpavtfgafvggnitllrgivyiiaqllgstvaclllkfvtndmavgvfslsagvg\n",
      " |      vtnalvfeivmtfglvytvyataidpkkgslgtiapiaigfivgani\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      To take a more annotation rich example,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> old = SeqIO.read(\"EMBL/TRBG361.embl\", \"embl\")\n",
      " |      >>> len(old.features)\n",
      " |      3\n",
      " |      >>> new = old.lower()\n",
      " |      >>> len(old.features) == len(new.features)\n",
      " |      True\n",
      " |      >>> old.annotations[\"organism\"] == new.annotations[\"organism\"]\n",
      " |      True\n",
      " |      >>> old.dbxrefs == new.dbxrefs\n",
      " |      True\n",
      " |  \n",
      " |  reverse_complement(self, id=False, name=False, description=False, features=True, annotations=False, letter_annotations=True, dbxrefs=False)\n",
      " |      Return new SeqRecord with reverse complement sequence.\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the reversed sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or True to keep that of the parent, or False to\n",
      " |      omit them. The default is to keep the original features (with the\n",
      " |      strand and locations adjusted).\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent,\n",
      " |      or False to omit them. The default is to keep the original\n",
      " |      annotations (with the letter annotations reversed).\n",
      " |      \n",
      " |      To show what happens to the pre-letter annotations, consider an\n",
      " |      example Solexa variant FASTQ file with a single entry, which we'll\n",
      " |      read in as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Now take the reverse complement, here we explicitly give a new\n",
      " |      identifier (the old identifier with a suffix):\n",
      " |      \n",
      " |      >>> rc_record = record.reverse_complement(id=record.id + \"_rc\")\n",
      " |      >>> print(\"%s %s\" % (rc_record.id, rc_record.seq))\n",
      " |      slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT\n",
      " |      \n",
      " |      Notice that the per-letter-annotations have also been reversed,\n",
      " |      although this may not be appropriate for all cases.\n",
      " |      \n",
      " |      >>> print(rc_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]\n",
      " |      \n",
      " |      Now for the features, we need a different example. Parsing a GenBank\n",
      " |      file is probably the easiest way to get an nice example with features\n",
      " |      in it...\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      >>> plasmid.seq\n",
      " |      Seq('GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG', IUPACAmbiguousDNA())\n",
      " |      >>> len(plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Now, let's take the reverse complement of this whole plasmid:\n",
      " |      \n",
      " |      >>> rc_plasmid = plasmid.reverse_complement(id=plasmid.id+\"_rc\")\n",
      " |      >>> print(\"%s %i\" % (rc_plasmid.id, len(rc_plasmid)))\n",
      " |      pBAD30_rc 4923\n",
      " |      >>> rc_plasmid.seq\n",
      " |      Seq('CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC', IUPACAmbiguousDNA())\n",
      " |      >>> len(rc_plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Let's compare the first CDS feature - it has gone from being the\n",
      " |      second feature (index 1) to the second last feature (index -2), its\n",
      " |      strand has changed, and the location switched round.\n",
      " |      \n",
      " |      >>> print(plasmid.features[1])\n",
      " |      type: CDS\n",
      " |      location: [1081:1960](-)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rc_plasmid.features[-2])\n",
      " |      type: CDS\n",
      " |      location: [2963:3842](+)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      You can check this new location, based on the length of the plasmid:\n",
      " |      \n",
      " |      >>> len(plasmid) - 1081\n",
      " |      3842\n",
      " |      >>> len(plasmid) - 1960\n",
      " |      2963\n",
      " |      \n",
      " |      Note that if the SeqFeature annotation includes any strand specific\n",
      " |      information (e.g. base changes for a SNP), this information is not\n",
      " |      amended, and would need correction after the reverse complement.\n",
      " |      \n",
      " |      Note trying to reverse complement a protein SeqRecord raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> protein_rec = SeqRecord(Seq(\"MAIVMGR\", IUPAC.protein), id=\"Test\")\n",
      " |      >>> protein_rec.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |      \n",
      " |      Also note you can reverse complement a SeqRecord using a MutableSeq:\n",
      " |      \n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Seq import MutableSeq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> rec = SeqRecord(MutableSeq(\"ACGT\", generic_dna), id=\"Test\")\n",
      " |      >>> rec.seq[0] = \"T\"\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      Test TCGT\n",
      " |      >>> rc = rec.reverse_complement(id=True)\n",
      " |      >>> print(\"%s %s\" % (rc.id, rc.seq))\n",
      " |      Test ACGA\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap=None, id=False, name=False, description=False, features=False, annotations=False, letter_annotations=False, dbxrefs=False)\n",
      " |      Return new SeqRecord with translated sequence.\n",
      " |      \n",
      " |      This calls the record's .seq.translate() method (which describes\n",
      " |      the translation related arguments, like table for the genetic code),\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the translated sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or False (default) to omit them.\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent\n",
      " |      (annotations only), or False (default) to omit them.\n",
      " |      \n",
      " |      e.g. Loading a FASTA gene and translating it,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> gene_record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> print(gene_record.format(\"fasta\"))\n",
      " |      >gi|3176602|gb|U78617.1|LOU78617 Lathyrus odoratus phytochrome A (PHYA) gene, partial cds\n",
      " |      CAGGCTGCGCGGTTTCTATTTATGAAGAACAAGGTCCGTATGATAGTTGATTGTCATGCA\n",
      " |      AAACATGTGAAGGTTCTTCAAGACGAAAAACTCCCATTTGATTTGACTCTGTGCGGTTCG\n",
      " |      ACCTTAAGAGCTCCACATAGTTGCCATTTGCAGTACATGGCTAACATGGATTCAATTGCT\n",
      " |      TCATTGGTTATGGCAGTGGTCGTCAATGACAGCGATGAAGATGGAGATAGCCGTGACGCA\n",
      " |      GTTCTACCACAAAAGAAAAAGAGACTTTGGGGTTTGGTAGTTTGTCATAACACTACTCCG\n",
      " |      AGGTTTGTT\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      And now translating the record, specifying the new ID and description:\n",
      " |      \n",
      " |      >>> protein_record = gene_record.translate(table=11,\n",
      " |      ...                                        id=\"phya\",\n",
      " |      ...                                        description=\"translation\")\n",
      " |      >>> print(protein_record.format(\"fasta\"))\n",
      " |      >phya translation\n",
      " |      QAARFLFMKNKVRMIVDCHAKHVKVLQDEKLPFDLTLCGSTLRAPHSCHLQYMANMDSIA\n",
      " |      SLVMAVVVNDSDEDGDSRDAVLPQKKKRLWGLVVCHNTTPRFV\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Return a copy of the record with an upper case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"acgtACGT\", generic_dna), id=\"Test\",\n",
      " |      ...                    description = \"Made up for this example\")\n",
      " |      >>> record.letter_annotations[\"phred_quality\"] = [1, 2, 3, 4, 5, 6, 7, 8]\n",
      " |      >>> print(record.upper().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      ACGTACGT\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Naturally, there is a matching lower method:\n",
      " |      \n",
      " |      >>> print(record.lower().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      acgtacgt\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  letter_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |      \n",
      " |      For example, this can hold quality scores used in FASTQ or QUAL files.\n",
      " |      Consider this example using Bio.SeqIO to read in an example Solexa\n",
      " |      variant FASTQ file as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      The letter_annotations get sliced automatically if you slice the\n",
      " |      parent SeqRecord, for example taking the last ten bases:\n",
      " |      \n",
      " |      >>> sub_record = record[-10:]\n",
      " |      >>> print(\"%s %s\" % (sub_record.id, sub_record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTNNNNNN\n",
      " |      >>> print(sub_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Any python sequence (i.e. list, tuple or string) can be recorded in\n",
      " |      the SeqRecord's letter_annotations dictionary as long as the length\n",
      " |      matches that of the SeqRecord's sequence.  e.g.\n",
      " |      \n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      1\n",
      " |      >>> sub_record.letter_annotations[\"dummy\"] = \"abcdefghij\"\n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      2\n",
      " |      \n",
      " |      You can delete entries from the letter_annotations dictionary as usual:\n",
      " |      \n",
      " |      >>> del sub_record.letter_annotations[\"solexa_quality\"]\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {'dummy': 'abcdefghij'}\n",
      " |      \n",
      " |      You can completely clear the dictionary easily as follows:\n",
      " |      \n",
      " |      >>> sub_record.letter_annotations = {}\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {}\n",
      " |      \n",
      " |      Note that if replacing the record's sequence with a sequence of a\n",
      " |      different length you must first clear the letter_annotations dict.\n",
      " |  \n",
      " |  seq\n",
      " |      The sequence itself, as a Seq or MutableSeq object.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Let us call help in the sequence object\n",
    "help(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two main attributes represent the id and the sequences.  Additionally, other attributes can contain additional information about our sequence. Let's print some attributes to see what they have."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sp|Q9FHQ6|UBQ9_ARATH\n",
      "sp|Q9FHQ6|UBQ9_ARATH\n",
      "sp|Q9FHQ6|UBQ9_ARATH Polyubiquitin 9 OS=Arabidopsis thaliana OX=3702 GN=UBQ9 PE=3 SV=1\n",
      "{}\n",
      "----------\n"
     ]
    }
   ],
   "source": [
    "# Sequence information\n",
    "print(sequence.id)\n",
    "print(sequence.name)\n",
    "print(sequence.description)\n",
    "print(sequence.annotations)\n",
    "print('----------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do we access the sequence?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSMQIHAKTLTEKTITIDVVSSDTINNVKAKIQDIEGIPLDQQRLIFSGKLLDDGRTLADYSIQKDSILHLALRLRGGMQIFVKTLTGKTITLEVESSDTIDNVKAKIQDKEGVPPDQQRLIFAGKQLDDGRTLADYNIQKESTLHLVLRLRGGMQIFVRTLTRKTIALEVESSDTTDNVKAKIQDKEGIPPDQQRLIFAGKQLEDGRTLADYNIQKESTLHLVLRLCGGMQIFVNTLTGKTITLEVESSDTIDNVKAKIQDKERIQPDQQRLIFAGEQLEDGYYTLADYNIQKESTLHLVLRLRGGECFGFIFLFLLCFNS\n",
      "\n",
      "ID: sp|Q9FHQ6|UBQ9_ARATH\n",
      "Name: sp|Q9FHQ6|UBQ9_ARATH\n",
      "Description: sp|Q9FHQ6|UBQ9_ARATH Polyubiquitin 9 OS=Arabidopsis thaliana OX=3702 GN=UBQ9 PE=3 SV=1\n",
      "Number of features: 0\n",
      "Seq('TEKTITIDVV', SingleLetterAlphabet())\n",
      "\n",
      "TEKTITIDVV\n",
      "----------\n",
      "<class 'Bio.SeqRecord.SeqRecord'>\n",
      "<class 'Bio.SeqRecord.SeqRecord'>\n",
      "----------\n",
      "<class 'Bio.Seq.Seq'>\n",
      "<class 'Bio.Seq.Seq'>\n",
      "----------\n"
     ]
    }
   ],
   "source": [
    "# Sequence itself\n",
    "print(sequence.seq)\n",
    "print()\n",
    "print(sequence[10:20]) # Slice directly on the SeqRecord object\n",
    "print()\n",
    "print(sequence.seq[10:20]) # Slice on the Seq object\n",
    "print('----------')\n",
    "# Print types\n",
    "print(type(sequence))\n",
    "print(type(sequence[10:20]))\n",
    "print('----------')\n",
    "print(type(sequence.seq))\n",
    "print(type(sequence.seq[10:20])) \n",
    "print('----------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can call help on the Seq type object to know more about it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Seq in module Bio.Seq object:\n",
      "\n",
      "class Seq(builtins.object)\n",
      " |  Read-only sequence object (essentially a string with an alphabet).\n",
      " |  \n",
      " |  Like normal python strings, our basic sequence object is immutable.\n",
      " |  This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      " |  Seq objects to be used as dictionary keys.\n",
      " |  \n",
      " |  The Seq object provides a number of string like methods (such as count,\n",
      " |  find, split and strip), which are alphabet aware where appropriate.\n",
      " |  \n",
      " |  In addition to the string like sequence, the Seq object has an alphabet\n",
      " |  property. This is an instance of an Alphabet class from Bio.Alphabet,\n",
      " |  for example generic DNA, or IUPAC DNA. This describes the type of molecule\n",
      " |  (e.g. RNA, DNA, protein) and may also indicate the expected symbols\n",
      " |  (letters).\n",
      " |  \n",
      " |  The Seq object also provides some biological methods, such as complement,\n",
      " |  reverse_complement, transcribe, back_transcribe and translate (which are\n",
      " |  not applicable to sequences with a protein alphabet).\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      If adding a string to a Seq, the alphabet is preserved:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> Seq(\"MELKI\", generic_protein) + \"LV\"\n",
      " |      Seq('MELKILV', ProteinAlphabet())\n",
      " |      \n",
      " |      When adding two Seq (like) objects, the alphabets are important.\n",
      " |      Consider this example:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet.IUPAC import unambiguous_dna, ambiguous_dna\n",
      " |      >>> unamb_dna_seq = Seq(\"ACGT\", unambiguous_dna)\n",
      " |      >>> ambig_dna_seq = Seq(\"ACRGT\", ambiguous_dna)\n",
      " |      >>> unamb_dna_seq\n",
      " |      Seq('ACGT', IUPACUnambiguousDNA())\n",
      " |      >>> ambig_dna_seq\n",
      " |      Seq('ACRGT', IUPACAmbiguousDNA())\n",
      " |      \n",
      " |      If we add the ambiguous and unambiguous IUPAC DNA alphabets, we get\n",
      " |      the more general ambiguous IUPAC DNA alphabet:\n",
      " |      \n",
      " |      >>> unamb_dna_seq + ambig_dna_seq\n",
      " |      Seq('ACGTACRGT', IUPACAmbiguousDNA())\n",
      " |      \n",
      " |      However, if the default generic alphabet is included, the result is\n",
      " |      a generic alphabet:\n",
      " |      \n",
      " |      >>> Seq(\"\") + ambig_dna_seq\n",
      " |      Seq('ACRGT')\n",
      " |      \n",
      " |      You can't add RNA and DNA sequences:\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import generic_dna, generic_rna\n",
      " |      >>> Seq(\"ACGT\", generic_dna) + Seq(\"ACGU\", generic_rna)\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      TypeError: Incompatible alphabets DNAAlphabet() and RNAAlphabet()\n",
      " |      \n",
      " |      You can't add nucleotide and protein sequences:\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import generic_dna, generic_protein\n",
      " |      >>> Seq(\"ACGT\", generic_dna) + Seq(\"MELKI\", generic_protein)\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      TypeError: Incompatible alphabets DNAAlphabet() and ProteinAlphabet()\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implement the 'in' keyword, like a python string.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna, generic_rna, generic_protein\n",
      " |      >>> my_dna = Seq(\"ATATGAAATTTGAAAA\", generic_dna)\n",
      " |      >>> \"AAA\" in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\") in my_dna\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\", generic_dna) in my_dna\n",
      " |      True\n",
      " |      \n",
      " |      Like other Seq methods, this will raise a type error if another Seq\n",
      " |      (or Seq like) object with an incompatible alphabet is used:\n",
      " |      \n",
      " |      >>> Seq(\"AAA\", generic_rna) in my_dna\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      TypeError: Incompatible alphabets DNAAlphabet() and RNAAlphabet()\n",
      " |      >>> Seq(\"AAA\", generic_protein) in my_dna\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      TypeError: Incompatible alphabets DNAAlphabet() and ProteinAlphabet()\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Compare the sequence to another sequence or a string (README).\n",
      " |      \n",
      " |      Historically comparing Seq objects has done Python object comparison.\n",
      " |      After considerable discussion (keeping in mind constraints of the\n",
      " |      Python language, hashes and dictionary support), Biopython now uses\n",
      " |      simple string comparison (with a warning about the change).\n",
      " |      \n",
      " |      Note that incompatible alphabets (e.g. DNA to RNA) will trigger a\n",
      " |      warning.\n",
      " |      \n",
      " |      During this transition period, please just do explicit comparisons:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> seq1 = Seq(\"ACGT\")\n",
      " |      >>> seq2 = Seq(\"ACGT\")\n",
      " |      >>> id(seq1) == id(seq2)\n",
      " |      False\n",
      " |      >>> str(seq1) == str(seq2)\n",
      " |      True\n",
      " |      \n",
      " |      The new behaviour is to use string-like equality:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> seq1 == seq2\n",
      " |      True\n",
      " |      >>> seq1 == \"ACGT\"\n",
      " |      True\n",
      " |      >>> seq1 == Seq(\"ACGT\", generic_dna)\n",
      " |      True\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Implement the greater-than or equal operand.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a subsequence of single letter, use my_seq[index].\n",
      " |      \n",
      " |      >>> my_seq = Seq('ACTCGACGTCG')\n",
      " |      >>> my_seq[5]\n",
      " |      'A'\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Implement the greater-than operand.\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Hash for comparison.\n",
      " |      \n",
      " |      See the __cmp__ documentation - this has changed from past\n",
      " |      versions of Biopython!\n",
      " |  \n",
      " |  __imul__(self, other)\n",
      " |      Multiply Seq in-place.\n",
      " |      \n",
      " |      Note although Seq is immutable, the in-place method is\n",
      " |      included to match the behaviour for regular Python strings.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> seq = Seq('ATG', generic_dna)\n",
      " |      >>> seq *= 2\n",
      " |      >>> seq\n",
      " |      Seq('ATGATG', DNAAlphabet())\n",
      " |  \n",
      " |  __init__(self, data, alphabet=Alphabet())\n",
      " |      Create a Seq object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - seq - Sequence, required (string)\n",
      " |       - alphabet - Optional argument, an Alphabet object from\n",
      " |         Bio.Alphabet\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      " |      the seq property.\n",
      " |      \n",
      " |      However, will often want to create your own Seq objects directly:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...              IUPAC.protein)\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      >>> print(my_seq)\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      >>> my_seq.alphabet\n",
      " |      IUPACProtein()\n",
      " |  \n",
      " |  __le__(self, other)\n",
      " |      Implement the less-than or equal operand.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence, use len(my_seq).\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Implement the less-than operand.\n",
      " |  \n",
      " |  __mul__(self, other)\n",
      " |      Multiply Seq by integer.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> Seq('ATG') * 2\n",
      " |      Seq('ATGATG')\n",
      " |      >>> Seq('ATG', generic_dna) * 2\n",
      " |      Seq('ATGATG', DNAAlphabet())\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Implement the not-equal operand.\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add a sequence on the left.\n",
      " |      \n",
      " |      If adding a string to a Seq, the alphabet is preserved:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> \"LV\" + Seq(\"MELKI\", generic_protein)\n",
      " |      Seq('LVMELKI', ProteinAlphabet())\n",
      " |      \n",
      " |      Adding two Seq (like) objects is handled via the __add__ method.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return (truncated) representation of the sequence for debugging.\n",
      " |  \n",
      " |  __rmul__(self, other)\n",
      " |      Multiply integer by Seq.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> 2 * Seq('ATG')\n",
      " |      Seq('ATGATG')\n",
      " |      >>> 2 * Seq('ATG', generic_dna)\n",
      " |      Seq('ATGATG', DNAAlphabet())\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return the full sequence as a python string, use str(my_seq).\n",
      " |      \n",
      " |      Note that Biopython 1.44 and earlier would give a truncated\n",
      " |      version of repr(my_seq) for str(my_seq).  If you are writing code\n",
      " |      which need to be backwards compatible with really old Biopython,\n",
      " |      you should continue to use my_seq.tostring() as follows::\n",
      " |      \n",
      " |          try:\n",
      " |              # The old way, removed in Biopython 1.73\n",
      " |              as_string = seq_obj.tostring()\n",
      " |          except AttributeError:\n",
      " |              # The new way, needs Biopython 1.45 or later.\n",
      " |              # Don't use this on Biopython 1.44 or older as truncates\n",
      " |              as_string = str(seq_obj)\n",
      " |  \n",
      " |  back_transcribe(self)\n",
      " |      Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\",\n",
      " |      ...                     IUPAC.unambiguous_rna)\n",
      " |      >>> messenger_rna\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG', IUPACUnambiguousRNA())\n",
      " |      >>> messenger_rna.back_transcribe()\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())\n",
      " |      \n",
      " |      Trying to back-transcribe a protein or DNA sequence raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      " |      >>> my_protein.back_transcribe()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins cannot be back transcribed!\n",
      " |  \n",
      " |  complement(self)\n",
      " |      Return the complement sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> my_dna = Seq(\"CCCCCGATAG\", IUPAC.unambiguous_dna)\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCGATAG', IUPACUnambiguousDNA())\n",
      " |      >>> my_dna.complement()\n",
      " |      Seq('GGGGGCTATC', IUPACUnambiguousDNA())\n",
      " |      \n",
      " |      You can of course used mixed case sequences,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> my_dna = Seq(\"CCCCCgatA-GD\", generic_dna)\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCgatA-GD', DNAAlphabet())\n",
      " |      >>> my_dna.complement()\n",
      " |      Seq('GGGGGctaT-CH', DNAAlphabet())\n",
      " |      \n",
      " |      Note in the above example, ambiguous character D denotes\n",
      " |      G, A or T so its complement is H (for C, T or A).\n",
      " |      \n",
      " |      Trying to complement a protein sequence raises an exception.\n",
      " |      \n",
      " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      " |      >>> my_protein.complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |  \n",
      " |  count(self, sub, start=0, end=9223372036854775807)\n",
      " |      Return a non-overlapping count, like that of a python string.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name,\n",
      " |      which does a non-overlapping count!\n",
      " |      \n",
      " |      For an overlapping search use the newer count_overlap() method.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count(\"A\"))\n",
      " |      5\n",
      " |      >>> print(my_seq.count(\"ATG\"))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> print(my_seq.count(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      \n",
      " |      HOWEVER, please note because python strings and Seq objects (and\n",
      " |      MutableSeq objects) do a non-overlapping search, this may not give\n",
      " |      the answer you expect:\n",
      " |      \n",
      " |      >>> \"AAAA\".count(\"AA\")\n",
      " |      2\n",
      " |      >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      " |      2\n",
      " |      \n",
      " |      An overlapping search, as implemented in .count_overlap(),\n",
      " |      would give the answer as three!\n",
      " |  \n",
      " |  count_overlap(self, sub, start=0, end=9223372036854775807)\n",
      " |      Return an overlapping count.\n",
      " |      \n",
      " |      For a non-overlapping search use the count() method.\n",
      " |      \n",
      " |      Returns an integer, the number of occurrences of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      Optional arguments start and end are interpreted as in slice\n",
      " |      notation.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      " |      3\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      " |      4\n",
      " |      >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      " |      1\n",
      " |      \n",
      " |      Where substrings do not overlap, should behave the same as\n",
      " |      the count() method:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"AAAATGA\")\n",
      " |      >>> print(my_seq.count_overlap(\"A\"))\n",
      " |      5\n",
      " |      >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"ATG\"))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      " |      True\n",
      " |      >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      " |      1\n",
      " |      >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      " |      True\n",
      " |      \n",
      " |      HOWEVER, do not use this method for such cases because the\n",
      " |      count() method is much for efficient.\n",
      " |  \n",
      " |  encode(self, encoding='utf-8', errors='strict')\n",
      " |      Return an encoded version of the sequence as a bytes object.\n",
      " |      \n",
      " |      The Seq object aims to match the interfact of a Python string.\n",
      " |      (This means on Python 3 it acts like a unicode string.)\n",
      " |      \n",
      " |      This is essentially to save you doing str(my_seq).encode() when\n",
      " |      you need a bytes string, for example for computing a hash:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> Seq(\"ACGT\").encode(\"ascii\")\n",
      " |      b'ACGT'\n",
      " |  \n",
      " |  endswith(self, suffix, start=0, end=9223372036854775807)\n",
      " |      Return True if the Seq ends with the given suffix, False otherwise.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Return True if the sequence ends with the specified suffix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      suffix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.endswith(\"UUG\")\n",
      " |      True\n",
      " |      >>> my_rna.endswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      " |      True\n",
      " |      >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      " |      True\n",
      " |  \n",
      " |  find(self, sub, start=0, end=9223372036854775807)\n",
      " |      Find method, like that of a python string.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Returns an integer, the index of the first occurrence of substring\n",
      " |      argument sub in the (sub)sequence given by [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"AUG\")\n",
      " |      3\n",
      " |  \n",
      " |  index(self, sub, start=0, end=9223372036854775807)\n",
      " |      Like find() but raise ValueError when the substring is not found.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.find(\"T\")\n",
      " |      -1\n",
      " |      >>> my_rna.index(\"T\")\n",
      " |      Traceback (most recent call last):\n",
      " |                 ...\n",
      " |      ValueError: substring not found...\n",
      " |  \n",
      " |  join(self, other)\n",
      " |      Return a merge of the sequences in other, spaced by the sequence from self.\n",
      " |      \n",
      " |      Accepts all Seq objects and Strings as objects to be concatenated with the spacer\n",
      " |      \n",
      " |      >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      " |      >>> concatenated\n",
      " |      Seq('AAANNNNNTTTNNNNNPPP')\n",
      " |      \n",
      " |      Throws error if other is not an iterable and if objects inside of the iterable\n",
      " |      are not Seq or String objects\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Return a lower case copy of the sequence.\n",
      " |      \n",
      " |      This will adjust the alphabet if required. Note that the IUPAC\n",
      " |      alphabets are upper case only, and thus a generic alphabet must be\n",
      " |      substituted.\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import Gapped, generic_dna\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"CGGTACGCTTATGTCACGTAG*AAAAAA\",\n",
      " |      ...          Gapped(IUPAC.unambiguous_dna, \"*\"))\n",
      " |      >>> my_seq\n",
      " |      Seq('CGGTACGCTTATGTCACGTAG*AAAAAA', Gapped(IUPACUnambiguousDNA(), '*'))\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('cggtacgcttatgtcacgtag*aaaaaa', Gapped(DNAAlphabet(), '*'))\n",
      " |      \n",
      " |      See also the upper method.\n",
      " |  \n",
      " |  lstrip(self, chars=None)\n",
      " |      Return a new Seq object with leading (left) end stripped.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Optional argument chars defines which characters to remove.  If\n",
      " |      omitted or None (default) then as for the python string method,\n",
      " |      this defaults to removing any white space.\n",
      " |      \n",
      " |      e.g. print(my_seq.lstrip(\"-\"))\n",
      " |      \n",
      " |      See also the strip and rstrip methods.\n",
      " |  \n",
      " |  reverse_complement(self)\n",
      " |      Return the reverse complement sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> my_dna = Seq(\"CCCCCGATAGNR\", IUPAC.ambiguous_dna)\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCGATAGNR', IUPACAmbiguousDNA())\n",
      " |      >>> my_dna.reverse_complement()\n",
      " |      Seq('YNCTATCGGGGG', IUPACAmbiguousDNA())\n",
      " |      \n",
      " |      Note in the above example, since R = G or A, its complement\n",
      " |      is Y (which denotes C or T).\n",
      " |      \n",
      " |      You can of course used mixed case sequences,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> my_dna = Seq(\"CCCCCgatA-G\", generic_dna)\n",
      " |      >>> my_dna\n",
      " |      Seq('CCCCCgatA-G', DNAAlphabet())\n",
      " |      >>> my_dna.reverse_complement()\n",
      " |      Seq('C-TatcGGGGG', DNAAlphabet())\n",
      " |      \n",
      " |      Trying to complement a protein sequence raises an exception:\n",
      " |      \n",
      " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      " |      >>> my_protein.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |  \n",
      " |  rfind(self, sub, start=0, end=9223372036854775807)\n",
      " |      Find from right method, like that of a python string.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Returns an integer, the index of the last (right most) occurrence of\n",
      " |      substring argument sub in the (sub)sequence given by [start:end].\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - sub - a string or another Seq object to look for\n",
      " |       - start - optional integer, slice start\n",
      " |       - end - optional integer, slice end\n",
      " |      \n",
      " |      Returns -1 if the subsequence is NOT found.\n",
      " |      \n",
      " |      e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.rfind(\"AUG\")\n",
      " |      15\n",
      " |  \n",
      " |  rindex(self, sub, start=0, end=9223372036854775807)\n",
      " |      Like rfind() but raise ValueError when the substring is not found.\n",
      " |  \n",
      " |  rsplit(self, sep=None, maxsplit=-1)\n",
      " |      Do a right split method, like that of a python string.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Return a list of the 'words' in the string (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done COUNTING FROM THE RIGHT.\n",
      " |      If maxsplit is omitted, all splits are made.\n",
      " |      \n",
      " |      Following the python string method, sep will by default be any\n",
      " |      white space (tabs, spaces, newlines) but this is unlikely to\n",
      " |      apply to biological sequences.\n",
      " |      \n",
      " |      e.g. print(my_seq.rsplit(\"*\",1))\n",
      " |      \n",
      " |      See also the split method.\n",
      " |  \n",
      " |  rstrip(self, chars=None)\n",
      " |      Return a new Seq object with trailing (right) end stripped.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Optional argument chars defines which characters to remove.  If\n",
      " |      omitted or None (default) then as for the python string method,\n",
      " |      this defaults to removing any white space.\n",
      " |      \n",
      " |      e.g. Removing a nucleotide sequence's polyadenylation (poly-A tail):\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"CGGTACGCTTATGTCACGTAGAAAAAA\", IUPAC.unambiguous_dna)\n",
      " |      >>> my_seq\n",
      " |      Seq('CGGTACGCTTATGTCACGTAGAAAAAA', IUPACUnambiguousDNA())\n",
      " |      >>> my_seq.rstrip(\"A\")\n",
      " |      Seq('CGGTACGCTTATGTCACGTAG', IUPACUnambiguousDNA())\n",
      " |      \n",
      " |      See also the strip and lstrip methods.\n",
      " |  \n",
      " |  split(self, sep=None, maxsplit=-1)\n",
      " |      Split method, like that of a python string.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Return a list of the 'words' in the string (as Seq objects),\n",
      " |      using sep as the delimiter string.  If maxsplit is given, at\n",
      " |      most maxsplit splits are done.  If maxsplit is omitted, all\n",
      " |      splits are made.\n",
      " |      \n",
      " |      Following the python string method, sep will by default be any\n",
      " |      white space (tabs, spaces, newlines) but this is unlikely to\n",
      " |      apply to biological sequences.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_aa = my_rna.translate()\n",
      " |      >>> my_aa\n",
      " |      Seq('VMAIVMGR*KGAR*L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      >>> for pep in my_aa.split(\"*\"):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      Seq('KGAR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      Seq('L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      >>> for pep in my_aa.split(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      Seq('KGAR*L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      \n",
      " |      See also the rsplit method:\n",
      " |      \n",
      " |      >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      " |      ...     pep\n",
      " |      Seq('VMAIVMGR*KGAR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      Seq('L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |  \n",
      " |  startswith(self, prefix, start=0, end=9223372036854775807)\n",
      " |      Return True if the Seq starts with the given prefix, False otherwise.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Return True if the sequence starts with the specified prefix\n",
      " |      (a string or another Seq object), False otherwise.\n",
      " |      With optional start, test sequence beginning at that position.\n",
      " |      With optional end, stop comparing sequence at that position.\n",
      " |      prefix can also be a tuple of strings to try.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      " |      >>> my_rna.startswith(\"GUC\")\n",
      " |      True\n",
      " |      >>> my_rna.startswith(\"AUG\")\n",
      " |      False\n",
      " |      >>> my_rna.startswith(\"AUG\", 3)\n",
      " |      True\n",
      " |      >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      " |      True\n",
      " |  \n",
      " |  strip(self, chars=None)\n",
      " |      Return a new Seq object with leading and trailing ends stripped.\n",
      " |      \n",
      " |      This behaves like the python string method of the same name.\n",
      " |      \n",
      " |      Optional argument chars defines which characters to remove.  If\n",
      " |      omitted or None (default) then as for the python string method,\n",
      " |      this defaults to removing any white space.\n",
      " |      \n",
      " |      e.g. print(my_seq.strip(\"-\"))\n",
      " |      \n",
      " |      See also the lstrip and rstrip methods.\n",
      " |  \n",
      " |  tomutable(self)\n",
      " |      Return the full sequence as a MutableSeq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\",\n",
      " |      ...              IUPAC.protein)\n",
      " |      >>> my_seq\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAAL', IUPACProtein())\n",
      " |      >>> my_seq.tomutable()\n",
      " |      MutableSeq('MKQHKAMIVALIVICITAVVAAL', IUPACProtein())\n",
      " |      \n",
      " |      Note that the alphabet is preserved.\n",
      " |  \n",
      " |  transcribe(self)\n",
      " |      Return the RNA sequence from a DNA sequence by creating a new Seq object.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\",\n",
      " |      ...                  IUPAC.unambiguous_dna)\n",
      " |      >>> coding_dna\n",
      " |      Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())\n",
      " |      >>> coding_dna.transcribe()\n",
      " |      Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG', IUPACUnambiguousRNA())\n",
      " |      \n",
      " |      Trying to transcribe a protein or RNA sequence raises an exception:\n",
      " |      \n",
      " |      >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      " |      >>> my_protein.transcribe()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins cannot be transcribed!\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap=None)\n",
      " |      Turn a nucleotide sequence into a protein sequence by creating a new Seq object.\n",
      " |      \n",
      " |      This method will translate DNA or RNA sequences, and those with a\n",
      " |      nucleotide or generic alphabet.  Trying to translate a protein\n",
      " |      sequence raises an exception.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - table - Which codon table to use?  This can be either a name\n",
      " |         (string), an NCBI identifier (integer), or a CodonTable\n",
      " |         object (useful for non-standard genetic codes).  This\n",
      " |         defaults to the \"Standard\" table.\n",
      " |       - stop_symbol - Single character string, what to use for\n",
      " |         terminators.  This defaults to the asterisk, \"*\".\n",
      " |       - to_stop - Boolean, defaults to False meaning do a full\n",
      " |         translation continuing on past any stop codons (translated as the\n",
      " |         specified stop_symbol).  If True, translation is terminated at\n",
      " |         the first in frame stop codon (and the stop_symbol is not\n",
      " |         appended to the returned protein sequence).\n",
      " |       - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      " |         this checks the sequence starts with a valid alternative start\n",
      " |         codon (which will be translated as methionine, M), that the\n",
      " |         sequence length is a multiple of three, and that there is a\n",
      " |         single in frame stop codon at the end (this will be excluded\n",
      " |         from the protein sequence, regardless of the to_stop option).\n",
      " |         If these tests fail, an exception is raised.\n",
      " |       - gap - Single character string to denote symbol used for gaps.\n",
      " |         It will try to guess the gap character from the alphabet.\n",
      " |      \n",
      " |      e.g. Using the standard table:\n",
      " |      \n",
      " |      >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      " |      >>> coding_dna.translate()\n",
      " |      Seq('VAIVMGR*KGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      >>> coding_dna.translate(stop_symbol=\"@\")\n",
      " |      Seq('VAIVMGR@KGAR@', HasStopCodon(ExtendedIUPACProtein(), '@'))\n",
      " |      >>> coding_dna.translate(to_stop=True)\n",
      " |      Seq('VAIVMGR', ExtendedIUPACProtein())\n",
      " |      \n",
      " |      Now using NCBI table 2, where TGA is not a stop codon:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2)\n",
      " |      Seq('VAIVMGRWKGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      " |      >>> coding_dna.translate(table=2, to_stop=True)\n",
      " |      Seq('VAIVMGRWKGAR', ExtendedIUPACProtein())\n",
      " |      \n",
      " |      In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      " |      this sequence could be a complete CDS:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=2, cds=True)\n",
      " |      Seq('MAIVMGRWKGAR', ExtendedIUPACProtein())\n",
      " |      \n",
      " |      It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      " |      and also the in frame stop codons:\n",
      " |      \n",
      " |      >>> coding_dna.translate(table=1, cds=True)\n",
      " |      Traceback (most recent call last):\n",
      " |          ...\n",
      " |      Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      " |      \n",
      " |      If the sequence has no in-frame stop codon, then the to_stop argument\n",
      " |      has no effect:\n",
      " |      \n",
      " |      >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      " |      >>> coding_dna2.translate()\n",
      " |      Seq('LAIVMGR', ExtendedIUPACProtein())\n",
      " |      >>> coding_dna2.translate(to_stop=True)\n",
      " |      Seq('LAIVMGR', ExtendedIUPACProtein())\n",
      " |      \n",
      " |      When translating gapped sequences, the gap character is inferred from\n",
      " |      the alphabet:\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import Gapped\n",
      " |      >>> coding_dna3 = Seq(\"GTG---GCCATT\", Gapped(IUPAC.unambiguous_dna))\n",
      " |      >>> coding_dna3.translate()\n",
      " |      Seq('V-AI', Gapped(ExtendedIUPACProtein(), '-'))\n",
      " |      \n",
      " |      It is possible to pass the gap character when the alphabet is missing:\n",
      " |      \n",
      " |      >>> coding_dna4 = Seq(\"GTG---GCCATT\")\n",
      " |      >>> coding_dna4.translate(gap='-')\n",
      " |      Seq('V-AI', Gapped(ExtendedIUPACProtein(), '-'))\n",
      " |      \n",
      " |      NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      " |      or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      " |      (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      " |      \n",
      " |      NOTE - This does NOT behave like the python string's translate\n",
      " |      method.  For that use str(my_seq).translate(...) instead.\n",
      " |  \n",
      " |  ungap(self, gap=None)\n",
      " |      Return a copy of the sequence without the gap character(s).\n",
      " |      \n",
      " |      The gap character can be specified in two ways - either as an explicit\n",
      " |      argument, or via the sequence's alphabet. For example:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\", generic_dna)\n",
      " |      >>> my_dna\n",
      " |      Seq('-ATA--TGAAAT-TTGAAAA', DNAAlphabet())\n",
      " |      >>> my_dna.ungap(\"-\")\n",
      " |      Seq('ATATGAAATTTGAAAA', DNAAlphabet())\n",
      " |      \n",
      " |      If the gap character is not given as an argument, it will be taken from\n",
      " |      the sequence's alphabet (if defined). Notice that the returned\n",
      " |      sequence's alphabet is adjusted since it no longer requires a gapped\n",
      " |      alphabet:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC, Gapped, HasStopCodon\n",
      " |      >>> my_pro = Seq(\"MVVLE=AD*\", HasStopCodon(Gapped(IUPAC.protein, \"=\")))\n",
      " |      >>> my_pro\n",
      " |      Seq('MVVLE=AD*', HasStopCodon(Gapped(IUPACProtein(), '='), '*'))\n",
      " |      >>> my_pro.ungap()\n",
      " |      Seq('MVVLEAD*', HasStopCodon(IUPACProtein(), '*'))\n",
      " |      \n",
      " |      Or, with a simpler gapped DNA example:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC, Gapped\n",
      " |      >>> my_seq = Seq(\"CGGGTAG=AAAAAA\", Gapped(IUPAC.unambiguous_dna, \"=\"))\n",
      " |      >>> my_seq\n",
      " |      Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
      " |      >>> my_seq.ungap()\n",
      " |      Seq('CGGGTAGAAAAAA', IUPACUnambiguousDNA())\n",
      " |      \n",
      " |      As long as it is consistent with the alphabet, although it is\n",
      " |      redundant, you can still supply the gap character as an argument to\n",
      " |      this method:\n",
      " |      \n",
      " |      >>> my_seq\n",
      " |      Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
      " |      >>> my_seq.ungap(\"=\")\n",
      " |      Seq('CGGGTAGAAAAAA', IUPACUnambiguousDNA())\n",
      " |      \n",
      " |      However, if the gap character given as the argument disagrees with that\n",
      " |      declared in the alphabet, an exception is raised:\n",
      " |      \n",
      " |      >>> my_seq\n",
      " |      Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
      " |      >>> my_seq.ungap(\"-\")\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Gap '-' does not match '=' from alphabet\n",
      " |      \n",
      " |      Finally, if a gap character is not supplied, and the alphabet does not\n",
      " |      define one, an exception is raised:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> my_dna = Seq(\"ATA--TGAAAT-TTGAAAA\", generic_dna)\n",
      " |      >>> my_dna\n",
      " |      Seq('ATA--TGAAAT-TTGAAAA', DNAAlphabet())\n",
      " |      >>> my_dna.ungap()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Gap character not given and not defined in alphabet\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Return an upper case copy of the sequence.\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import HasStopCodon, generic_protein\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> my_seq = Seq(\"VHLTPeeK*\", HasStopCodon(generic_protein))\n",
      " |      >>> my_seq\n",
      " |      Seq('VHLTPeeK*', HasStopCodon(ProteinAlphabet(), '*'))\n",
      " |      >>> my_seq.lower()\n",
      " |      Seq('vhltpeek*', HasStopCodon(ProteinAlphabet(), '*'))\n",
      " |      >>> my_seq.upper()\n",
      " |      Seq('VHLTPEEK*', HasStopCodon(ProteinAlphabet(), '*'))\n",
      " |      \n",
      " |      This will adjust the alphabet if required. See also the lower method.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(sequence.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To learn more about how Biopython handles sequence files, please visit:\n",
    "\n",
    "[Biopython SeqIO](https://biopython.org/wiki/SeqIO)\n",
    "\n",
    "### Align sequences with biopython\n",
    "\n",
    "Biopython contains methods to align a pair of sequences, either globally or locally.\n",
    "Let's pick up two sequences and align them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Seq.Seq'>\n",
      "<class 'Bio.Seq.Seq'>\n"
     ]
    }
   ],
   "source": [
    "# These are two random DNA sequences of 20bp length\n",
    "seq1 = Seq('GGGCGGCCTGGCCGGAGCAC')\n",
    "seq2 = Seq('TCCTCGGCGATGTGTGTAAA')\n",
    "\n",
    "print(type(seq1))\n",
    "print(type(seq2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will employ the Bio.pairwise2 module to align the two sequences globally:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "201\n"
     ]
    }
   ],
   "source": [
    "# align and print the number of different alignments\n",
    "alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "print(len(alignments))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The method generates many alignments that have similar scores. We can access all of them by iterating the alignment object. We do so and print the resulting five first alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('GGGCGG-CCT-GGCCGGA-GC-----AC--', '------TCCTCGG-C-GATG-TGTGTA-AA', 10.0, 0, 30)\n",
      "('GGGCGGCCT-GGCCGGA-GC-----AC--', '-----TCCTCGG-C-GATG-TGTGTA-AA', 10.0, 0, 29)\n",
      "('GGGCGGCCT-GGCCGGA-GC-----AC--', '----T-CCTCGG-C-GATG-TGTGTA-AA', 10.0, 0, 29)\n",
      "('GGGCGGCCT-GGCCGGA-GC-----AC--', '---T--CCTCGG-C-GATG-TGTGTA-AA', 10.0, 0, 29)\n",
      "('GGGCGGCCT-GGCCGGA-GC-----AC--', '--T---CCTCGG-C-GATG-TGTGTA-AA', 10.0, 0, 29)\n"
     ]
    }
   ],
   "source": [
    "# Print first 5 alignments\n",
    "for a in alignments[:5]:\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can improve the visualization by using the \"format_alignment\" function of the Biopython pairwise2 method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.pairwise2 import format_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pretty print first alignment:\n",
      "GGGCGG-CCT-GGCCGGA-GC-----AC--\n",
      "       ||| || | || |      |   \n",
      "------TCCTCGG-C-GATG-TGTGTA-AA\n",
      "  Score=10\n",
      "\n",
      "Print alignment sequences:\n",
      "GGGCGG-CCT-GGCCGGA-GC-----AC--\n",
      "------TCCTCGG-C-GATG-TGTGTA-AA\n"
     ]
    }
   ],
   "source": [
    "print('Pretty print first alignment:')\n",
    "print(format_alignment(*alignments[0]))\n",
    "print('Print alignment sequences:')\n",
    "# We access each sequence with indexes\n",
    "print(alignments[0][0]) # first sequence with index 0\n",
    "print(alignments[0][1]) # second sequence with index 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that for each alignment generated, we can access the individual strings of each sequence by just using indexing over the alignment object.\n",
    "\n",
    "Now using this, let's calculate the percentage of identity for each alignment. For this, we create a function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculatePID(alignment):\n",
    "    \n",
    "    # Get the two sequences\n",
    "    seq1 = alignment[0]\n",
    "    seq2 = alignment[1]\n",
    "    \n",
    "    count = 0 # Variable containing the count of matching characters\n",
    "    # iterate\n",
    "    for j in range(len(seq1)):\n",
    "        p1 = seq1[j] # Position i of first alignment \n",
    "        p2 = seq2[j] # Position i of second alignment\n",
    "    \n",
    "        # Only count when we don't have '-' at any position.\n",
    "        if p1 != '-' and p2 != '-':\n",
    "            # Only when the sequences match\n",
    "            if p1 == p1:\n",
    "                count += 1\n",
    "            \n",
    "    # Get length of each sequence\n",
    "    l1 = len([p for p in seq1 if p != '-'])\n",
    "    l2 = len([p for p in seq2 if p != '-'])\n",
    "    \n",
    "    # the PID uses the length of the shortest sequence to convert to percentage\n",
    "    pid = count/min(l1, l2)\n",
    "    return pid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using this function, let's calculate the percentage of identity (PID) of an alignment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment 0 PID:\n",
      "0.5\n"
     ]
    }
   ],
   "source": [
    "print('Alignment 0 PID:')\n",
    "print(calculatePID(alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate a PID matrix\n",
    "\n",
    "We now move into comparing sequences in an all vs. all fashion. This type of comparison is advantageous if we want to find groups of related sequences. To this end, We need to compare every sequence to each other. We will need a matrix to store the comparison data (NumPy offers a convenient way of storaging matrix data). Let's have a look at how a matrix of data looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print a 5x5 matrix filled with zeros\n",
    "M = np.zeros((5,5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each (i,j) element of the matrix contains a value that corresponds to the comparison of elements i and j. \n",
    "\n",
    "We now read a set of N sequences and store them in a dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the sequences in the file\n",
    "sequences = SeqIO.parse('input/apolipoproteinCI_sequences.fasta', 'fasta')\n",
    "# Convert iterator to dictionary\n",
    "sequences = {s.id : s for s in sequences}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will calculate the percentage of identity between each pair of sequences\n",
    "and store the M matrix values. First, we initialize a matrix with the correct shape and filled with zeros to keep all the values (NxN)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48, 48)\n"
     ]
    }
   ],
   "source": [
    "# Initialize the NxN matrix  matrix\n",
    "N = len(sequences)\n",
    "M = np.zeros((N, N))\n",
    "print(M.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we compare all the sequences to each other. To iterate the elements to compare, we use a double for loop; one loop for i and one for j elements:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,s1 in enumerate(sequences):\n",
    "    for j, s2 in enumerate(sequences):\n",
    "        seq1 = sequences[s1]\n",
    "        seq2 = sequences[s2]\n",
    "        if i == j: # the same element always has a PID=1.0\n",
    "            M[i][j] = 1.0\n",
    "        if j > i : # We compare only half matrix\n",
    "            alignment = pairwise2.align.globalxx(sequences[s1], sequences[s2])\n",
    "            M[i][j] = calculatePID(alignment[0])\n",
    "#             M[j][i] = M[i][j] # Uncomment this to fill the lower half matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's visualize how our matrix looks like using matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.image.AxesImage at 0x1524e8216a20>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQEAAAECCAYAAAD+eGJTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAfT0lEQVR4nO2da4yc53Xf/2dm57r3XS6p5Z2yKUEXOxJMS1btto4cIaocWE6KFBHilEDdKh/sxgaCxnJbNPA3FQUcF6j7QWgc0XDixInUSJEVywJtIlGtSKEu1I0SKepCkbvc++x17vv0A2eVfZ5zljPiZXdHz/8HLHafs+d958w7s2ff58y5iHMOhJB4SWy0AYSQjYVOgJDIoRMgJHLoBAiJHDoBQiKHToCQyFl3JyAid4rIGyLypojct96P/0EQke+JyLiIvLJKNiAiT4rIycb3/o200UJEdonIz0XkuIi8KiJfa8g3te0ikhWRZ0XkWMPubzXkm9ruFUQkKSIviMhjjXVb2L2uTkBEkgC+C+BfAbgewD0icv162vABeRDAnYHsPgCHnXP7ARxurDcbNQC/75y7DsCnAHylcZ03u+1lALc7534JwE0A7hSRT2Hz273C1wAcX7VuD7udc+v2BeA2AE+sWn8TwDfX04aLsHkvgFdWrd8AMNz4eRjAGxttYwvP4REAd7ST7QDyAJ4HcGs72A1gJ87/od8O4LF2eq+s93ZgB4D3Vq3PNGTtxDbn3CgANL5v3WB7LoiI7AVwM4Bn0Aa2N26pXwQwDuBJ51xb2A3gOwD+AMDyKlk72L3uTkAMGfOWrxAi0gXgIQBfd87NbbQ9reCcqzvnbsL5/6y3iMiNG21TM0Tk1wCMO+ee22hbLob1dgJnAOxatd4JYGSdbbhUxkRkGAAa38c32B4TEUnhvAP4U+fcww1xW9gOAM65AoAjOB+T2ex2fxrAF0TkHQB/DuB2EfkBNr/dANbfCfwjgP0isk9E0gB+C8Cj62zDpfIogIONnw/i/H57UyEiAuCPARx3zn171a82te0iMiQifY2fcwB+BcDr2OR2O+e+6Zzb6Zzbi/Pv6Z85576ETW73+2xAAOUuACcAnALwXzY6KNLE1h8CGAVQxfm7mC8DGMT5ANDJxveBjbbTsPszOL/NegnAi42vuza77QA+DuCFht2vAPhvDfmmtjt4Dp/FPwUG28JuaRhLCIkUZgwSEjl0AoREDp0AIZFDJ0BI5NAJEBI5G+IEROTejXjcy0G72t6udgPta3u72L1RdwJtcXHWoF1tb1e7gfa1vS3sviQn0E69AQghNhedLNToDXAC50tUz+B8SvA9zrnX1jpmy0DS7d2VwsRUHUODSQDASC3r6WQTVXVc3WlfVXVJb71s+LOEV9DVkIn/fCvLHUonJXUlyzTsmp+uoXugA+XllNKpGXamE7WmjxfaBADO+bVWy0btVVpqShaea8WmxZkKOvvTAIClelodF5JN6HMvtnCc9Vyq9aSh5782znh+yYZOZbaIdG/OfLyOhH6Na8v6dagHslyHfp+VjdemFTLGtaq7BMqFIjJ9uTVtsq5V+B6ydJaXrTo8n1TSfw8Xz82jXCiaB17csz7PLQDedM69BQAi8ucA7gawphPYuyuFZ5/Y5cn+cOIGb31D7ow6brrWpWSj1T5vXTL+KPOJipYly9763eIWpbM9U1Cyj2bPees3S1cpnalqp5Ltyk576/dKA0onl2z+hiwaf4DhuQGgK1ny1uPVHqVzrLBTyUInc233mNI5OrVbyUI6U/qan53tVbLurP86VI0/kq60PlfIYHZRyaZK+nUoFH0ncuOWUaXzzrx+baw/wpDdXTNKNl/NeOvJon4P541rNVPKBTr6vTFXyiqZBHbu7Pbfw0f+/V+pY1a4lO1AS70BROReETkqIkcnpvR/WELIxnIpTqCl3gDOuQeccweccwdWtgCEkM3DpWwHPnBvgBMv5fGr22/yZE+MvOitf/fMbS09+M1dp7113fBnC3V927StY9Zbz3ToW8ek6H3midKwtz6Qf1vpvFPRW4ur034J+Wwtr3SG03r7Ed7G39R5Wum8WdqmZNdm/dvcHSl9qzpV0bemz4zt8dbFmt5eWfvvVMK/uyuU9N69VNHn6s2Vmurs69HbnaWavy2ytlJD2QUlG531r2cYqwHsW/9QZsU3Jkr6ei4HMZ2kce225eaVLMS65oWifl8ng3/J08GWKIyJrOZS7gQ+DL0BCImei74TcM7VROSrAJ4AkATwPefcq5fNMkLIunAp2wE45x4H8PhlsoUQsgFckhO4HOgYwdNK5z+du1nJZuv+3vpsuU/pWJ/fTlb9/dtcTe+vrL3hdTk/3PGT2Y8pnWJd72tfTw431bE+WixU/b31bE3vta2PDX9WvU7Jmp0bAAbz/kdtO/KzSuelye1Nz91pfKyXTet9+2LFtz38iAsASnX99gxlo0v6I9B8h/HRcMaXzRrXwNo3lwJZuNcHgD4jhhR+/PfuvJ47En6MCOiPUzMp/R5eNuwM832yOf+aW9d3BRYQERI5dAKERA6dACGRQydASORseGAwJAwUAsATIy8o2X9479PeuryskzjuHnxRyVJB0c1PCzrAd01W55Wfqfh55fNVHVC0ClC6gloFq4AoaQbF/ADiWLlb6dze/7qSvbzo1wVYSVRhgg+gk1Jem9GJSBcKLq1QqunnV65qmYgfuMoZwcOa069pNXid7SIcfdxS2Q9EloxkKMv2ECt4uFjTAdowgHmhZJ3VWNdB2WAEJ11w/mzSf59fqAaCdwKERA6dACGRQydASOTQCRASOZsuMGhhBwv/n7f+/pyu4PuHhY80PXdPR1HJTpV1UCxsWnJT93tKZ3/mnJKVnH/c0wsfVTr7MhNK9nrCzzT8Z90nlc7hwvVKdku3X91oVRE+Xvi4kp0q+9cv7GwEANkOnb0WdrApG8G1et3ouNThH7dQ0tlzH+md1McFWaBpM8ipZcmEn7EXdjYCWnt+82Vtp9WlKLRzvqKPs7sw+deqL6i2BIDFsg5EhucKq0CtTMf3j13zN4SQKKATICRy6AQIiZy2iAlYNOtQBAB/aOztw067YTIPoDsZA0B30MCzbnRXO1vVlWJh9+TwPIC9NwyTinZ36C47VpJRWF051DGndJaNrshh55uc0dGmN61tDzG7+Bpxgr29/vM5NaNjOpad4d52S0Z3EZqu6KrMgc4lb92f1rGgMEEL0K9NJq/jBoMZ3ex0tuInkw3ltM4W47iRlF9F2J3S789Ch05U6wySjMIGrFaHohV4J0BI5NAJEBI5dAKERA6dACGR07aBwRA7oUgHC3+65Ad/jszrdlzX5HTSz/Gi31rLGuX1Vn1IyfZmp7y11SbMmp6UCEY4PDGvqx0/36+f32slf/7LW5WtSmfSCJztCqbojBd11WLFqNQMq/oySR04s5gIJvIMd+sAZqGir1UYqLOCgNZxYfJT1Qg6Wm3Ww+cTPt+1ZK2ce9xoVZ4LJg5ZLdYSRt5PJWiFHgZMO4zkqPfPt+ZvCCFRQCdASOTQCRASOXQChETOhyYwaNFKsPDJOR3USWDtIMoKvUb14ZwR9Aur1awMxXBcOqCzFq3WZRYpaT752aooC7PzuoxMtfElHSwMsTIGrTl8YYAvDIQCQM0I3oVYgTPr+c0FVYqVTn2clVVXDs5vPT9rHuJSEAi0WrOFLcAAIJz2ED4+YM92CJ9z+H65UGM43gkQEjl0AoREDp0AIZHzoY4JWOjqQ93O/Ktnb1WycG+/lNDJQpmE3hvO1Pxkls/36ASfl0u7lOzqtN9t6K2KTkT68YyOeQyl/Zn3dWNffUO3bqn+kxE/acqqUNzfpzsghZwsaDv7sjp+Mlv2K+EKRR1P2dc3pWRLQXtva19tVWVu7/FlYXINAEwt5pXsxiH/Wr0zN6h0JhI66SdMDrISrSxZ2LnohgGduPbC+A4lC5Ohzi75szmtVvcr8E6AkMihEyAkcugECIkcOgFCIie6wGCInVD0jJJ9a8Jv793bsaR0LMJ24n82fZvSySR0cGu+5gfO+ozH+2zvcSV7u+xXDW5LheknwF+P3axkn9hyxlu/taADYKcXdPu0ECsIuFg1gqhBK+8v7XlW2zmqX5swqci6dmNGBWR43O5O3Yrdamc2ttTjP57RljxrJBD1Z/zXy7p2C1Xdhrw347dwO7PYp3TyxrzCMNHpuh4/oPi8EbRegXcChEQOnQAhkUMnQEjkNHUCIvI9ERkXkVdWyQZE5EkROdn43nyzSAjZlLQSGHwQwP8C8P1VsvsAHHbO3S8i9zXW37j85m0MrVQffnv6aqVjtRybrulsspBkC1WLS8v63CdKw0oWVh+eqQw0PTcAnJzzM/0+Ofiu0nlqXM92rC37/0es9mLTRibe1m6//dXDIzpYabEczHtYNK65VUUYzgJcqOnjCmWdtRieK5/SFXzh3EEAKNabzwK0qhbngkxKKxA5b8xtDGc7Hp+7yrfHaGG3QtM7Aefc3wEIJ1/cDeBQ4+dDAL7Y7DyEkM3JxcYEtjnnRgGg8V13sySEtAVXPE9ARO4FcC8AZKFvCwkhG8vFOoExERl2zo2KyDCA8bUUnXMPAHgAAHpk4EINTjY1rcw+/G5BVwN+JO1fmsmqjhEMp3VCTxgD2J8bUzqjFZ1I8svdr3nrqbp+vJG8Pu4fRvd466fdPqUT7v8Bva+1WmtXqvptNr3k779LFX3cJ7a/p2SZhL/3HSvpxKAdnQUlO1rY7a0Htunkq1eMGMvWLj92MbGoW5yHFXyAjl1YcZFwXiAALJb91936g6nWjIrE4Bp3BMlY1mu3wsVuBx4FcLDx80EAj1zkeQghG0wrHxH+EMDTAK4VkTMi8mUA9wO4Q0ROArijsSaEtCFNtwPOuXvW+NXnLrMthJANgBmDhERO9FWEF0ursw8fnPM/PV2o60SPktMvQ5hA1J3Q1Xlv1K5SsveqfvXf/rRuT2Ulygx2+oEyK5CUSjZvZ24FyVIpnfCSC2TJhA6BWS2xChU/wDac00HV2apO+ukIkmnOlXqUTngNAD1nsDujk4W60yUlUy3Ac621OM8H5+8ykpOWUvr1SwTn6kn7LeOTnEVICFkLOgFCIodOgJDIoRMgJHIYGLyMtBIsfLe8RekMdcwr2WzND4AtG/76qvSckg0k/Qy3xWUdiOzq0MGmMHC0LEaAL6EDg+FcAzsApQN1uZTf7soKkpWM7MMw4LZYMwKtxnFhwHJHVmcVjhmzFkO7UkYrMWv2QWinNcPQIhU8njXD0JIlgpcrnP+oX81Vx7ZkGSHkQwudACGRQydASOQwJnCFaaX68Nmy3i8Odfj7/WNLe5TOSFlXA84EsYR8Uu//J8u6Ei6cDRgmm1g6gN7D7u3R8wOv7taysPPOsfHtSqcno5NwynX/LWvN8wsr+ABgT7/fYvy5qd1Kp1Rr/udw3YCu5tya0TGdsLPPTqOycV9+UslOTPkdnqpp/X/aSsiq1Hy9MCZxofJd3gkQEjl0AoREDp0AIZFDJ0BI5DAwuM60Wn34P6Z3eOs9GR1ECoOAgG5DlhadyDJS0gHFcy0kqfQbcwbDKrtrOnWnuV9M6fbs4aw8q4pwd17PC5wLkoOG0gtKp2y0135hyr+eVnWe1RY8bPm9Oxc23tbzJgHg1IKfFDac1dWOwykdLAxbh1t2po1qTivZqlV4J0BI5NAJEBI5dAKERA6dACGRI86t3yiAHhlwtwr7k7ZCGCz8+ugBpRNm3QHAtXk/MGjNOXhzcUjJ3pj026CFASoAWG7hrRJWswF2kLG+HFQDFnU14I4BYx5D1X/OvUZWYRisBIC5kp/taNm0WNJtuzJBGzTruJ6szq4MmS3qbMtrBnVAMXwdOpI64LenVwcnw8DgO7P+DMpXf+9BLJ4YNYsJeSdASOTQCRASOXQChEQOk4U2Kbr68KjS+b2RTypZIujsc1Pnu0rnjJEs1J/3E4GqRsvxfEpXO4atya2klSljDl+4j142Hm84rzsnLQVdg7pSej++7PS5qnU/TrC7WycinZ7vV7KwzXrCiAlYMwVnSv5zHurUOrmkvp5hy3Hrmr81M6hkYbJVGDc4aXSFWoF3AoREDp0AIZFDJ0BI5NAJEBI5DAy2CXb14T8q2bcmrvfW4xU9c2+uoluAb+/0E3OmSroF2VxFJ/TUg4BeZ9poZ25UCIaBx+EeHQScKOpEp7CdmBUELBotx+dLvu21rtb+/y1W/AQiq4KvUtfJSa3YNFvRCURhNs98WV9zq73YcpD0FwZorSSnFXgnQEjk0AkQEjl0AoREDp0AIZHDwGAb00qrsocWdGCwYLQle23G75OfMebrWRV7IdYcwKWylqWDtl3TRW3T7h6d1bdU8wN1/Zklfe6krgZ8b8bPkuxN6VZpE0kdiMwH7b2s52cRZhGeWdBZmlb2YdjOrM+YYThlXKtkcK6RhV5vfaHgJe8ECIkcOgFCIodOgJDIaRoTEJFdAL4P4CoAywAecM79TxEZAPAXAPYCeAfAv3HO6U0cWVdamX3404JONgmTSazkli5jPmErOvNpvUcPW2lbMwXTRuVbLeHvmQdSujrP6izUk/fjGTVDZyin25cvVP1kHev59aZ1rGS+heOsisuw7XmHaJ1w/iOg4wt9GT/mcaGW5K3cCdQA/L5z7joAnwLwFRG5HsB9AA475/YDONxYE0LajKZOwDk36px7vvHzPIDjAHYAuBvAoYbaIQBfvFJGEkKuHB8oJiAiewHcDOAZANucc6PAeUcBYOsax9wrIkdF5GgVzW8nCSHrS8tOQES6ADwE4OvOOV3tsQbOuQeccweccwdS0MUQhJCNpaVkIRFJ4bwD+FPn3MMN8ZiIDDvnRkVkGIAeQkc2HDuh6Gkl+0LxTm+dTuhkoVv73276eM/N7lGyLXkdhOvP+kk+Y0vdSmdbRv+vqWf8/1sZw07ruMmsXxVpzQY8a7Rd+0Sv357txOJVSmdfTs+JPFP2W5Ul0Fpr/zAweNeWl5XOT6ZuVLIwgBheg2eM6/RPtjVBRATAHwM47pz79qpfPQrgYOPngwAeaXYuQsjmo5U7gU8D+B0AL4vIyudN/xnA/QB+JCJfBnAawG9eGRMJIVeSpk7AOfcUdK+DFThOiJA2hxmDhEQOZxESADqz8M7XP690Zsu6HVa52nxHeXX/lJKdnPLnId4wdE7pWHMAisHjdWd0O7NwFgIA7O/z5/5NlHTF4FBWZwweG9/urcN2aoDdGi3f4ds1uqirOTNGq7KeoFJzdF4fN9Sp7SzX/euyu8tP3v3xwUcxeXySswgJIRo6AUIih06AkMhhZyECwKo+/LHS+cxLv6Fk+bTf+aZutMM+Nb1FycLZh6NLeu9rzQusBS3Gt2R0FWHBaKl+fHqbt76mX+e2TZV1m/WuIOYQJvMAwBZjFuFc1Y+f9BhVhFbVYhgH2dGtk5renhlQsrCKsDuY0RheN+/YNX9DCIkCOgFCIodOgJDIoRMgJHIYGCQmVvXhUyMPK9npmh/c+pOZW5TOz8euUbIb+0e99UvT25VOGFwDgGrQOnt7TgfOrFZlW4MEm7BFN2AnGT18/Q+89Z/MflzpvLagbV+s+i3VrusdUzojRSOBKGg5/kd7H1I6B0u/o2QhpSB5yAporsA7AUIih06AkMihEyAkcugECIkcBgZJy7Qy+zBp9MkPZxoAwGDaD9SVavqtuDWv5/BZwbsQq8d+bdm3YbGmZyFYLAVVtuVlPY+h1dZhIV0pY0aDEQwNSRrPz6pubBXeCRASOXQChEQOnQAhkcOYALkkWpl9+PaSriK8IXfGW/9NXbfRtuYFDmT8VuWzVV0xaLVLX076yTITRd1ZaNCoBvyrOT856Ld7jyqdZzK7lOx/z37WW3+m54TSea+qqwHfmRv01t8v6OQrq/owTAZaCmIea6cK8U6AkOihEyAkcugECIkcOgFCIoeBQXJZsROKnlKyr5691Vvv7i0onWJNJ+bMV/2htpmkDgKG7b4B4JcH3/DWf3b6k0pnT35ayX42ea23fnF+p9LZmdW2h9WAj09/TOkMpJeU7KZBP2B6+Ny1SufWoXeUbK7mJxlNBoFPVhESQtaEToCQyKETICRy6AQIiRwGBskVxw4WPuOtf+nZe5TOx7aOKpnq558qKZ1UQs/4+5tzfuZfb0Yf91Jhh5Jd3T3prT/X95rSKdT1vILnsNtb/+62I0rnF0v7tZ1nfTtvHNDX4O3FQSULsyv7Mv5cB6vycAXeCRASOXQChEQOnQAhkcOYANkQwjjBsZEfKp3bjv1rJStV/ASizoxODLLmIe7r8ROBFmoZpbMjr9uXH5v04wQvTOhkoe6M7hAUJjH9x1d0zMMZdv7zHae89cszup35tvy8koUxgY4gLsIqQkLImtAJEBI5dAKERE5TJyAiWRF5VkSOicirIvKthnxARJ4UkZON7/3NzkUI2Xy0EhgsA7jdObcgIikAT4nI3wL4DQCHnXP3i8h9AO4D8I0raCv5EGMlFD09oufwnar6rbX+vni10vmXubeU7ODr/vy+T255V+lYSTh9WT/pxmoTbs1DLARtz9JGAtNdW19WsodHb2563FxFtyUPW473Z/127Vbb9xWa3gm486xc+VTjywG4G8ChhvwQgC82OxchZPPRUkxARJIi8iKAcQBPOueeAbDNOTcKAI3vW6+cmYSQK0VLTsA5V3fO3QRgJ4BbRES3hl0DEblXRI6KyNEq9K0UIWRj+UCfDjjnCgCOALgTwJiIDANA4/v4Gsc84Jw74Jw7kIJO0CCEbCxNA4MiMgSg6pwriEgOwK8A+O8AHgVwEMD9je+PXElDSXy0MvvwyJL+P9aX0LIwi3C6qiv/rDmAw0EW4dnFPqUzmNHzCqbL/vmtVmnWXMOFiv+PsjOlMyKtVmGLVX/OQM3516BqzHBYoZVPB4YBHBKRJM7fOfzIOfeYiDwN4Eci8mUApwH8ZgvnIoRsMpo6AefcSwBuNuRTAD53JYwihKwfzBgkJHJYRUjailZmHz44t1fJhnL+vr03VVQ65az+c0iJ35Hnoz0TSmdXdkbJxoo93np3l6FT7VGyXMpP8rmu75zSebUwrGRdaf+Tt3Bmo9VtaQXeCRASOXQChEQOnQAhkUMnQEjkMDBI2ppWEooA4MHTflVfsku34C7V9Z9DmHRjVfVVnU7E6Un7gceZcl7p3NZ7SsmOit+q/PSirtAfyi4oWaHiP7/lCzYU8+GdACGRQydASOTQCRASOXQChEQOA4PkQ4cVLDwy8tfe+rOv6EZYrbQcmyh2KZ2Fqi6Rnyz6VYRhRh8A/OXIJ5RsMOtnNpaM6sOtWT13ICSc2Vh3a/+/550AIZFDJ0BI5NAJEBI5jAmQKAjjBGGMAAC+ePJXlSwdzBSsGB16tmd0y/H+oIpvrNitdAYyOulnqeZ3CAofHwCen9ylZGG3od3dftViUnRy1Aq8EyAkcugECIkcOgFCIodOgJDIYWCQRIldffiEkoXBwm15nagzbyQLhQlE+Q7dOjxjBP125Are+mxRtzi/cWBUyTrEr248PnuVt64tM1mIELIGdAKERA6dACGRQydASOQwMEhIg1aChf91/GNKZ7TUq2Slul/9NxhkEALAlNFyLGxD9kd7H1I6/+7EbytZrqN6wXVCnDrm/d+t+RtCSBTQCRASOXQChEQOYwKEXIBWZh9+t6Cr+v5i9oC3/rfbf6F0XlrarWQ/P7ffWz+ycKPS6U2XbGNXsRhUI4ZVhqvhnQAhkUMnQEjk0AkQEjl0AoREDgODhHwAWp19+PzcHm/9bmWL0tmXmVCy1/J+9d+xeR10HMwsKlk4e7Ar5bc4TyXYXowQsgZ0AoREDp0AIZHTshMQkaSIvCAijzXWAyLypIicbHzXg9QJIZueDxIY/BqA4wB6Guv7ABx2zt0vIvc11t+4zPYRsumxg4VPeeu73rhL6eztmlay5WBm4MnCkNLpMIJ8yUA2lPNnGqxdQ9jinYCI7ATweQD/Z5X4bgCHGj8fAqAnPBJCNj2tbge+A+APAKx2N9ucc6MA0Pi+1TpQRO4VkaMicrQKPZmVELKxNHUCIvJrAMadc89dzAM45x5wzh1wzh1IQXdlJYRsLK3EBD4N4AsicheALIAeEfkBgDERGXbOjYrIMIDxK2koIe2Erj58XOl8Z2avkp2c82MAB4ZOK53jhauULIwJdASzB9euIWzhTsA5903n3E7n3F4AvwXgZ865LwF4FMDBhtpBAI80OxchZPNxKXkC9wO4Q0ROArijsSaEtBkfqHbAOXcEwJHGz1MAPnf5TSKErCfMGCQkclhFSMg60Gr14d+e89uJnVnSswg7U/qj9jDJKJyFWGd7MULIWtAJEBI5dAKERA6dACGRw8AgIRuEHSx8zFt/4eSdSmdbdl7J+lL+rMPjc35WobtAziDvBAiJHDoBQiKHToCQyGFMgJBNhK4+/InS+Rcv/7qSJcTvHZRPVXyFC7QW4p0AIZFDJ0BI5NAJEBI5dAKERA4Dg4RsYqyEor8b+b9K9mql6K2/cuIeb73MKkJCyFrQCRASOXQChEQOnQAhkcPAICFtRiutyiTIILzQ4AHeCRASOXQChEQOnQAhkcOYACEfAsI4wc9H/KmAt2QLax7LOwFCIodOgJDIoRMgJHLoBAiJHHHuAn2HLveDiUwAeBfAFgCT6/bAl5d2tb1d7Qba1/bNZPce59yQ9Yt1dQLvP6jIUefcgXV/4MtAu9rernYD7Wt7u9jN7QAhkUMnQEjkbJQTeGCDHvdy0K62t6vdQPva3hZ2b0hMgBCyeeB2gJDIoRMgJHLoBAiJHDoBQiKHToCQyPn/5KQUyFopmswAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.matshow(M)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Wrapping up\n",
    "\n",
    "In this second practice session, we learned:\n",
    "\n",
    "- How to read multiple sequences from a file and store them as a dictionary\n",
    "- How to write a Python function for repetitive tasks\n",
    "- How to read sequences with Biopython \n",
    "- How to do a pairwise alignment of two sequences with Biopython\n",
    "- How to calculate a PID matrix"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
